<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyzing ChIP-seq data • OneLinerOmics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing ChIP-seq data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">OneLinerOmics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ChIPseqAnalysis.html">Analyzing ChIP-seq data</a>
    </li>
    <li>
      <a href="../articles/RNAseqAnalysis.html">Analyzing RNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/j-andrews7/OneLinerOmics">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Analyzing ChIP-seq data</h1>
                        <h4 class="author">Jared Andrews</h4>
            <address class="author_afil">
      Washington University School of Medicine, Department of Laboratory and Genomic Medicine<br><a class="author_email" href="mailto:#"></a><a href="mailto:jared.andrews07@gmail.com" class="email">jared.andrews07@gmail.com</a>
      </address>
                  
            <h4 class="date">Revised: November 15th, 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/j-andrews7/OneLinerOmics/blob/master/vignettes/ChIPseqAnalysis.Rmd"><code>vignettes/ChIPseqAnalysis.Rmd</code></a></small>
      <div class="hidden name"><code>ChIPseqAnalysis.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This vignette walks through ChIP-seq analysis of the B cell lymphoma ChIP-seq data included in the future Andrews et al. (2019, your favorite journal with an impact factor &gt; 30) paper. It uses the convenience functions included in this <a href="https://github.com/j-andrews7/AndrewsBCellLymphoma">package</a> to streamline the process.</p>
<p>Several command line tools as well as python are also utilized - each code block should indicate whether it is <code>bash</code>, <code>R</code>, or <code>python</code> code. If your system is set up properly, you can still <a href="https://bookdown.org/yihui/rmarkdown/language-engines.html">run <em>many of</em> these code chunks in Rmarkdown files within RStudio</a>, assuming you have all of the software installed.</p>
<p>Some of this code is best ran on a computing cluster, as it takes a significant amount of time locally. Such instances will have <code>- cluster</code> appended to the language statement at the beginning of the code block.</p>
<p>Required Software:</p>
<ul>
<li>R (3.6+)</li>
<li>python (ideally one <a href="https://docs.conda.io/projects/conda/en/latest/index.html">conda environment</a> with python2.7, another with python3)</li>
<li><a href="https://bitbucket.org/young_computation/rose/src/master/">ROSE</a></li>
<li><a href="http://www.htslib.org/download/">samtools</a></li>
<li><a href="https://bedtools.readthedocs.io/en/latest/">bedtools</a></li>
<li>
<a href="https://github.com/taoliu/MACS">MACS</a> - I used v2.1.0.20150420 but more recent ones are available now.</li>
<li><a href="https://github.com/ucscGenomeBrowser/kent">kentUtils</a></li>
<li><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">bowtie2</a></li>
</ul>
</div>
<div id="faqsuggested-reading" class="section level1">
<h1 class="hasAnchor">
<a href="#faqsuggested-reading" class="anchor"></a>FAQ/Suggested Reading</h1>
<p>That is, if you ask me one of the following questions and haven’t read/tried these first, I’m going to tell you to read/try them.</p>
<ul>
<li>“What does this error during alignment mean?”
<ul>
<li>Read the <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">bowtie2 manual</a>.</li>
</ul>
</li>
<li>“Why do I have so few peaks after peakcalling?”
<ul>
<li>Your data is probably garbage. This can be confirmed visually in a genome browser (no clear peaks == garbage) or by viewing the fraction of reads in peaks/cross-correlation values during QC. You can play with the <code>-m</code> and <code>-q</code> parameters for MACS2 to adjust the sensitivity, but usually it just means your IP didn’t work all that well.</li>
</ul>
</li>
<li>“Why won’t ROSE work?”
<ul>
<li>Do you have R, samtools, and python 2.7 installed?</li>
<li>Are you calling the command from the directory where you downloaded ROSE? Adding it to <code>PATH</code> does not seem to work.</li>
</ul>
</li>
<li>“What are these <code>mcfork</code> errors about from ChIPQC/DiffBind?”
<ul>
<li>Parallelization failure, especially if you’re on Windows/WSL. Running <code>register(SerialParam())</code> will usually fix this.</li>
</ul>
</li>
<li>“Your paths don’t make any sense! Are you some sorta dingus?”
<ul>
<li>Well, yeah, but much of the alignment, peak calling, etc., code is copy-pasted with minor editing merely to demonstrate the process and serve as a record for which parameters were used. It may require a smidgen of effort to adapt to your directory structure. The main purpose of this vignette is to show how the package works, which should be much more clear.</li>
</ul>
</li>
</ul>
<div id="alignment-indexing-and-blacklisted-read-removal" class="section level2">
<h2 class="hasAnchor">
<a href="#alignment-indexing-and-blacklisted-read-removal" class="anchor"></a>Alignment, Indexing, and Blacklisted Read Removal</h2>
<p>In order to generate tracks, perform peak calling, and compare signal, we need to align all of our files to the genome. This process will also sort and removed blacklisted reads from the resulting <code>BAM</code> file, then index it. It requires your genome sequence in <code>FASTA</code> format, which you can easily download. Using a different genome version (or organism) is easy - just swap out the version (hg19 -&gt; hg38) or look for <a href="https://hgdownload.soe.ucsc.edu/downloads.html">your organism</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" title="1"><span class="co"># bash</span></a>
<a class="sourceLine" id="cb1-2" title="2"><span class="fu">wget</span> http://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/hg19.fa.gz</a>
<a class="sourceLine" id="cb1-3" title="3"><span class="fu">gunzip</span> hg19.fa.gz</a>
<a class="sourceLine" id="cb1-4" title="4"></a>
<a class="sourceLine" id="cb1-5" title="5"><span class="co"># Grab chromosome sizes, need these for blacklisting.</span></a>
<a class="sourceLine" id="cb1-6" title="6"><span class="fu">wget</span> http://hgdownload.cse.ucsc.edu/goldenpath/hg19/bigZips/hg19.chrom.sizes</a>
<a class="sourceLine" id="cb1-7" title="7"><span class="fu">mkdir</span> ./ChIP_Seq/ref</a>
<a class="sourceLine" id="cb1-8" title="8"><span class="fu">mv</span> hg19.chrom.sizes ./ChIP_Seq/ref/hg19.chrom.sizes</a></code></pre></div>
<p><code>bowtie2</code> also has to build an index for the genome prior to use.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" title="1"><span class="co"># bash</span></a>
<a class="sourceLine" id="cb2-2" title="2"><span class="ex">bowtie-build</span> hg19.fa hg19</a></code></pre></div>
<p>Blacklisted regions are those that have high artificial signal in pretty much every ChIP experiment - usually due to being extremely repetitive or near centromeres. <a href="https://www.encodeproject.org/annotations/ENCSR636HFF/">ENCODE</a> identified these regions, so we can remove all such reads so they don’t bias our normalization/differential binding analyses later on. A (probably quicker) alternative whould be to remove peaks that overlap these regions.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" title="1"><span class="co"># bash</span></a>
<a class="sourceLine" id="cb3-2" title="2"><span class="fu">wget</span> https://www.encodeproject.org/files/ENCFF001TDO/@@download/ENCFF001TDO.bed.gz</a>
<a class="sourceLine" id="cb3-3" title="3"><span class="fu">gunzip</span> ENCFF001TDO.bed.gz</a>
<a class="sourceLine" id="cb3-4" title="4"><span class="fu">mv</span> ENCFF001TDO.bed ./ChIP_Seq/ref/ENCODE_blacklists.bed</a>
<a class="sourceLine" id="cb3-5" title="5"></a>
<a class="sourceLine" id="cb3-6" title="6"><span class="co"># Due to how samtools view works, we create a file of regions we WANT to keep rather than exclude.</span></a>
<a class="sourceLine" id="cb3-7" title="7"><span class="ex">bedtools</span> complement -i ./ChIP_Seq/ref/ENCODE_blacklists.bed \</a>
<a class="sourceLine" id="cb3-8" title="8">-g ./ChIP_Seq/ref/hg19.chrom.sizes <span class="op">&gt;</span> ./ChIP_Seq/ref/hg19_blacklist_regions_removed.bed</a></code></pre></div>
<p>Now we can perform the actual alignment.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" title="1"><span class="co"># bash - cluster</span></a>
<a class="sourceLine" id="cb4-2" title="2"></a>
<a class="sourceLine" id="cb4-3" title="3"><span class="co"># Directory containing bowtie indices, blacklist, etc.</span></a>
<a class="sourceLine" id="cb4-4" title="4"><span class="va">genome=</span><span class="st">"./ChIP_Seq/ref/hg19"</span></a>
<a class="sourceLine" id="cb4-5" title="5"><span class="co"># Base directory containing reads and where additional directories will be created.</span></a>
<a class="sourceLine" id="cb4-6" title="6"><span class="va">base_dir=</span><span class="st">"./ChIP_Seq/READS/"</span></a>
<a class="sourceLine" id="cb4-7" title="7"></a>
<a class="sourceLine" id="cb4-8" title="8"><span class="bu">cd</span> <span class="st">"</span><span class="va">$base_dir</span><span class="st">"</span></a>
<a class="sourceLine" id="cb4-9" title="9"><span class="fu">mkdir</span> BAMS</a>
<a class="sourceLine" id="cb4-10" title="10"></a>
<a class="sourceLine" id="cb4-11" title="11"><span class="co"># Alignment.</span></a>
<a class="sourceLine" id="cb4-12" title="12"><span class="kw"><a href="https://rdrr.io/r/base/Control.html">for</a></span> <span class="ex">f</span> in ./ChIP_Seq/READS/*.fq.gz<span class="kw">;</span> <span class="kw">do</span></a>
<a class="sourceLine" id="cb4-13" title="13">  <span class="bu">echo</span> Aligning <span class="st">"</span><span class="va">$f</span><span class="st">"</span></a>
<a class="sourceLine" id="cb4-14" title="14">  <span class="ex">bowtie2</span> -p 6 -x ./ChIP_Seq/ref/hg19 -U <span class="st">"</span><span class="va">$f</span><span class="st">"</span> <span class="kw"><a href="https://rdrr.io/r/base/Logic.html">|</a></span> <span class="ex">samtools</span> view -@ 8 -b -S -u - \</a>
<a class="sourceLine" id="cb4-15" title="15">  <span class="kw"><a href="https://rdrr.io/r/base/Logic.html">|</a></span> <span class="ex">samtools</span> sort -@ 8 -m 5G -o ./ChIP_Seq/BAMS/<span class="st">"</span><span class="va">${f%</span>.*<span class="va">}</span><span class="st">"</span>.sorted.bam -<span class="kw">;</span></a>
<a class="sourceLine" id="cb4-16" title="16">  <span class="ex">samtools</span> index ./ChIP_Seq/BAMS/<span class="st">"</span><span class="va">${f%</span>.*<span class="va">}</span><span class="st">"</span>.sorted.bam</a>
<a class="sourceLine" id="cb4-17" title="17"><span class="kw">done</span></a>
<a class="sourceLine" id="cb4-18" title="18"></a>
<a class="sourceLine" id="cb4-19" title="19"><span class="co"># Removal of blacklist reads.</span></a>
<a class="sourceLine" id="cb4-20" title="20"><span class="kw"><a href="https://rdrr.io/r/base/Control.html">for</a></span> <span class="ex">f</span> in ./ChIP_Seq/BAMS/*.bam<span class="kw">;</span> <span class="kw">do</span></a>
<a class="sourceLine" id="cb4-21" title="21">    <span class="bu">echo</span> Removing blacklisted reads from <span class="st">"</span><span class="va">$f</span><span class="st">"</span></a>
<a class="sourceLine" id="cb4-22" title="22">    <span class="va">base=${f##</span>*/<span class="va">}</span></a>
<a class="sourceLine" id="cb4-23" title="23">    <span class="ex">samtools</span> view -@ 8 -b -t ./ChIP_Seq/Ref/hg19.fa \</a>
<a class="sourceLine" id="cb4-24" title="24">    -L ./ChIP_Seq/Ref/hg19_blacklist_regions_removed.bed \</a>
<a class="sourceLine" id="cb4-25" title="25">    -o ./ChIP_Seq/BAMS/<span class="st">"</span><span class="va">${base%</span>.*<span class="va">}</span><span class="st">"</span>.BL_removed.bam <span class="st">"</span><span class="va">$f</span><span class="st">"</span><span class="kw">;</span></a>
<a class="sourceLine" id="cb4-26" title="26">    <span class="ex">samtools</span> index ./ChIP_Seq/BAMS/<span class="st">"</span><span class="va">${base%</span>.*<span class="va">}</span><span class="st">"</span>.BL_removed.bam<span class="kw">;</span></a>
<a class="sourceLine" id="cb4-27" title="27"><span class="kw">done</span></a></code></pre></div>
</div>
<div id="peak-calling" class="section level2">
<h2 class="hasAnchor">
<a href="#peak-calling" class="anchor"></a>Peak Calling</h2>
<p>Another essential step is peak calling. Tons of peak callers exist, some are better at broad peak calling, some are made for TFs, etc etc. <a href="https://github.com/taoliu/MACS">MACS</a> is one of the most common and most well-regarded peak callers out there. It generally does pretty well even if you botch the options, and the author has recently began developing it again. The script below is meant to run on a batch of files that have a corresponding input/IgG sample within the same folder.</p>
<p>Here I use a q-value of 0.01, as this should help reduce peaks called in samples with high background. You may adjust them as you see fit.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb5-1" title="1"><span class="co"># bash - cluster</span></a>
<a class="sourceLine" id="cb5-2" title="2"></a>
<a class="sourceLine" id="cb5-3" title="3"><span class="co"># This example is for H3K27AC samples, hence the wildcard use to capture that.</span></a>
<a class="sourceLine" id="cb5-4" title="4"></a>
<a class="sourceLine" id="cb5-5" title="5"><span class="kw"><a href="https://rdrr.io/r/base/Control.html">for</a></span> <span class="ex">f</span> in ./ChIP_Seq/BAMs/*C.sorted.BL_removed.bam<span class="kw">;</span> <span class="kw">do</span></a>
<a class="sourceLine" id="cb5-6" title="6">    <span class="va">base=${f##</span>*/<span class="va">}</span></a>
<a class="sourceLine" id="cb5-7" title="7">    <span class="ex">macs2</span> callpeak -t <span class="st">"</span><span class="va">$f</span><span class="st">"</span> -c ./ChIP_Seq/BAMs/*INPUT.sorted.BL_removed.bam -n <span class="st">"</span><span class="va">$base</span><span class="st">"</span> \</a>
<a class="sourceLine" id="cb5-8" title="8">    --outdir ./ChIP_Seq/MACS/H3K27AC --tempdir ./ChIP_Seq/scratch/MACS2 -q 0.01 \</a>
<a class="sourceLine" id="cb5-9" title="9">    --nomodel --shiftsize=150<span class="kw">;</span></a>
<a class="sourceLine" id="cb5-10" title="10"><span class="kw">done</span></a></code></pre></div>
<div id="pre-filter" class="section level3">
<h3 class="hasAnchor">
<a href="#pre-filter" class="anchor"></a>Pre-filter</h3>
<p>I also only care about autosomal peaks, so I filter those out. I also have several marks, so I organized the resulting files into directories.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb6-1" title="1"><span class="co"># bash</span></a>
<a class="sourceLine" id="cb6-2" title="2"><span class="va">marks=(</span>FAIRE H3AC H3K27AC H3K4ME1<span class="va">)</span></a>
<a class="sourceLine" id="cb6-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/Control.html">for</a></span> <span class="ex">m</span> in <span class="st">"</span><span class="va">${marks[@]}</span><span class="st">"</span><span class="kw">;</span> <span class="kw">do</span></a>
<a class="sourceLine" id="cb6-4" title="4">    <span class="kw"><a href="https://rdrr.io/r/base/Control.html">for</a></span> <span class="ex">f</span> in ./ChIP_Seq/MACS2/<span class="st">"</span><span class="va">$m</span><span class="st">"</span>/PEAKS_NARROW/*.narrowPeak<span class="kw">;</span> <span class="kw">do</span></a>
<a class="sourceLine" id="cb6-5" title="5">        <span class="fu">sed</span> <span class="st">'/_g\|chrM\|uN\|rand\|chrY\|chrX\|chr23/d'</span> <span class="st">"</span><span class="va">$f</span><span class="st">"</span> <span class="op">&gt;</span> <span class="st">"</span><span class="va">$f</span><span class="st">"</span>.clean</a>
<a class="sourceLine" id="cb6-6" title="6">        <span class="ex">rename.ul</span> .narrowPeak.clean .clean.narrowPeak <span class="st">"</span><span class="va">$f</span><span class="st">"</span>.clean</a>
<a class="sourceLine" id="cb6-7" title="7">    <span class="kw">done</span></a>
<a class="sourceLine" id="cb6-8" title="8"><span class="kw">done</span></a></code></pre></div>
</div>
</div>
<div id="super-enhancer-calling-with-rose" class="section level2">
<h2 class="hasAnchor">
<a href="#super-enhancer-calling-with-rose" class="anchor"></a>Super Enhancer Calling with ROSE</h2>
<p><a href="http://younglab.wi.mit.edu/super_enhancer_code.html">ROSE</a> was the original tool made for SE calling, and it’s still used today. Some other tools do the same thing, but they all take the same approach, so might as well stick with the OG, yeah?</p>
<p>First, we need our peaks in a <code>gff</code> format.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb7-1" title="1"><span class="co"># python</span></a>
<a class="sourceLine" id="cb7-2" title="2"></a>
<a class="sourceLine" id="cb7-3" title="3"><span class="im">import</span> glob</a>
<a class="sourceLine" id="cb7-4" title="4"><span class="co"># Store file names</span></a>
<a class="sourceLine" id="cb7-5" title="5">marks <span class="op">=</span> [<span class="st">"FAIRE"</span>, <span class="st">"H3AC"</span>, <span class="st">"H3K27AC"</span>, <span class="st">"H3K4ME1"</span>]</a>
<a class="sourceLine" id="cb7-6" title="6"></a>
<a class="sourceLine" id="cb7-7" title="7"><span class="cf">for</span> m <span class="kw"><a href="https://rdrr.io/r/base/Control.html">in</a></span> marks:</a>
<a class="sourceLine" id="cb7-8" title="8">  in_files <span class="op">=</span> glob.glob(<span class="st">"./ChIP_Seq/MACS2/"</span> <span class="op">+</span> m <span class="op">+</span> <span class="st">"/PEAKS_NARROW/*.narrowPeak"</span>)</a>
<a class="sourceLine" id="cb7-9" title="9">  </a>
<a class="sourceLine" id="cb7-10" title="10">  <span class="cf">for</span> n <span class="kw"><a href="https://rdrr.io/r/base/Control.html">in</a></span> in_files:</a>
<a class="sourceLine" id="cb7-11" title="11">    base <span class="op">=</span> <span class="st">"."</span>.join(input_file.split(<span class="st">"."</span>)[<span class="dv">0</span>:<span class="op">-</span><span class="dv">1</span>])</a>
<a class="sourceLine" id="cb7-12" title="12">    output_file <span class="op">=</span> <span class="st">"./ChIP_Seq/MACS2/"</span> <span class="op">+</span> m <span class="op">+</span> <span class="st">"/PEAKS_NARROW/"</span> <span class="op">+</span> base <span class="op">+</span> <span class="st">".gff"</span></a>
<a class="sourceLine" id="cb7-13" title="13">    </a>
<a class="sourceLine" id="cb7-14" title="14">    <span class="cf">with</span> <span class="bu">open</span>(n) <span class="im">as</span> f:</a>
<a class="sourceLine" id="cb7-15" title="15">    </a>
<a class="sourceLine" id="cb7-16" title="16">        <span class="co"># Open output file, "w" to make it writable</span></a>
<a class="sourceLine" id="cb7-17" title="17">        output_f <span class="op">=</span> <span class="bu">open</span>(output_file, <span class="st">"w"</span>)</a>
<a class="sourceLine" id="cb7-18" title="18">        new_ID <span class="op">=</span> <span class="dv">1</span></a>
<a class="sourceLine" id="cb7-19" title="19">    </a>
<a class="sourceLine" id="cb7-20" title="20">        <span class="cf">for</span> line <span class="kw"><a href="https://rdrr.io/r/base/Control.html">in</a></span> f:</a>
<a class="sourceLine" id="cb7-21" title="21">            line <span class="op">=</span> line.strip().split()</a>
<a class="sourceLine" id="cb7-22" title="22">    </a>
<a class="sourceLine" id="cb7-23" title="23">            <span class="co"># Handle if bed file already has an ID that should be retained.</span></a>
<a class="sourceLine" id="cb7-24" title="24">            <span class="cf">if</span> <span class="bu">len</span>(line) <span class="op">&gt;</span> <span class="dv">3</span>:</a>
<a class="sourceLine" id="cb7-25" title="25">                <span class="cf">if</span> line[<span class="dv">3</span>] <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span> <span class="kw">and</span> line[<span class="dv">3</span>] <span class="kw">is</span> <span class="kw">not</span> <span class="st">"."</span>:</a>
<a class="sourceLine" id="cb7-26" title="26">                    ID <span class="op">=</span> line[<span class="dv">3</span>]</a>
<a class="sourceLine" id="cb7-27" title="27">                <span class="cf">else</span>:</a>
<a class="sourceLine" id="cb7-28" title="28">                    ID <span class="op">=</span> new_ID</a>
<a class="sourceLine" id="cb7-29" title="29">            <span class="cf">else</span>:</a>
<a class="sourceLine" id="cb7-30" title="30">                ID <span class="op">=</span> new_ID</a>
<a class="sourceLine" id="cb7-31" title="31">    </a>
<a class="sourceLine" id="cb7-32" title="32">            <span class="bu">print</span>(line[<span class="dv">0</span>], ID, <span class="st">"PEAK"</span>, <span class="bu">str</span>(<span class="bu">int</span>(line[<span class="dv">1</span>]) <span class="op">+</span> <span class="dv">1</span>), line[<span class="dv">2</span>],</a>
<a class="sourceLine" id="cb7-33" title="33">                  <span class="st">"."</span>, <span class="st">"."</span>, <span class="st">"."</span>, ID, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="bu">file</span><span class="op">=</span>output_f)</a>
<a class="sourceLine" id="cb7-34" title="34">    </a>
<a class="sourceLine" id="cb7-35" title="35">            new_ID <span class="op">+=</span> <span class="dv">1</span></a>
<a class="sourceLine" id="cb7-36" title="36">    </a>
<a class="sourceLine" id="cb7-37" title="37">        output_f.close()</a></code></pre></div>
<p>Easy enough to use, just have to feed it the <code>BAM</code> file for both your sample and the input along with the peaks. Requires <code>samtools</code> and <code>R</code> on your <code>PATH</code>, so be sure both of those can be run. It also has to be run from the <code>rose</code> folder wherever you cloned it - trying to add the <code>rose</code> directory to your <code>PATH</code> really doesn’t work. These are pretty standard options. <code>-t</code> defines what you want to consider the promoter of a gene, in this case we’re using +/- 2500 bp. It excludes peaks in these regions from the stitching process.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb8-1" title="1"><span class="co"># bash</span></a>
<a class="sourceLine" id="cb8-2" title="2"><span class="va">base=</span>/mnt/f/BCellRework/ChIP_Seq</a>
<a class="sourceLine" id="cb8-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/Control.html">for</a></span> <span class="ex">f</span> in <span class="st">"</span><span class="va">$base</span><span class="st">"</span>/BAMs/K27AC/TS*K27AC.sorted.BL_removed.bam<span class="kw">;</span> <span class="kw">do</span></a>
<a class="sourceLine" id="cb8-4" title="4">    <span class="va">samp=${f##</span>*/<span class="va">}</span>;</a>
<a class="sourceLine" id="cb8-5" title="5">    <span class="va">samp=${samp%</span>.*<span class="va">}</span>;</a>
<a class="sourceLine" id="cb8-6" title="6">    <span class="va">short=${samp%%</span>_*<span class="va">}</span>;</a>
<a class="sourceLine" id="cb8-7" title="7">    <span class="bu">echo</span> <span class="st">"processing </span><span class="va">$samp</span><span class="st">"</span><span class="kw">;</span></a>
<a class="sourceLine" id="cb8-8" title="8">    <span class="ex">python</span> ROSE_main.py -g hg19 -r <span class="st">"</span><span class="va">$f</span><span class="st">"</span> -t 2500 -c <span class="st">"</span><span class="va">$base</span><span class="st">"</span>/BAMs/INPUT/<span class="st">"</span><span class="va">$short</span><span class="st">"</span>_INPUT.sorted.bam \</a>
<a class="sourceLine" id="cb8-9" title="9">    -o ../RESULTS/<span class="st">"</span><span class="va">$short</span><span class="st">"</span> -i <span class="st">"</span><span class="va">$base</span><span class="st">"</span>/ROSE/PEAKS_GFF/<span class="st">"</span><span class="va">$short</span><span class="st">"</span>_peaks.clean.gff<span class="kw">;</span></a>
<a class="sourceLine" id="cb8-10" title="10"><span class="kw">done</span></a></code></pre></div>
</div>
<div id="making-tracks" class="section level2">
<h2 class="hasAnchor">
<a href="#making-tracks" class="anchor"></a>Making Tracks</h2>
<p>Ultimately, we want to be able to visualize the results as well, so we can create at least semi-normalized tracks that account for read depth to make them more visually comparable. Always take track visualizations with a grain of salt - use them as representations of what you’ve found through more thorough statistical analyses, never try to use them to make claims or you’re gonna have a bad time.</p>
<p>I use <code>deepTools</code> to generate RPKM-normalized signal bigwig generation, and ignore chrM, X, and Y for these purposes. It can be installed with a simple <code>pip install deeptools</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb9-1" title="1"><span class="co"># bash</span></a>
<a class="sourceLine" id="cb9-2" title="2"><span class="fu">mkdir</span> ./ChIP_Seq/TRACKS</a>
<a class="sourceLine" id="cb9-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/Control.html">for</a></span> <span class="ex">f</span> in ./ChIP_Seq/BAMs/*BL_removed.bam<span class="kw">;</span> <span class="kw">do</span> </a>
<a class="sourceLine" id="cb9-4" title="4">  <span class="bu">echo</span> <span class="st">"</span><span class="va">$f</span><span class="st">"</span><span class="kw">;</span> </a>
<a class="sourceLine" id="cb9-5" title="5">  <span class="ex">bamCoverage</span> --bam <span class="st">"</span><span class="va">$f</span><span class="st">"</span> -o ./ChIP_Seq/TRACKS/<span class="st">"</span><span class="va">$f</span><span class="st">"</span>.rpkm.bw --binSize 10 \</a>
<a class="sourceLine" id="cb9-6" title="6">  --ignoreForNormalization chrX chrM chrY --extendReads 150 \</a>
<a class="sourceLine" id="cb9-7" title="7">  --normalizeUsing RPKM -p max/2<span class="kw">;</span> </a>
<a class="sourceLine" id="cb9-8" title="8"><span class="kw">done</span></a></code></pre></div>
<p>For the peaks, look up the <code>bigNarrowPeak</code> format and convert to it with <code>bedToBigBed</code>. You will have to fetch the hg19 chromosome sizes first with UCSC’s <code>fetchChromSizes</code> utility. The <a href="https://github.com/ucscGenomeBrowser/kent"><code>kentUtils</code> tools</a> will give you everything you need for this. You might have to change the score in the 5th column for a few lines by running something like:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb10-1" title="1"><span class="co"># bash</span></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/Control.html">for</a></span> <span class="ex">f</span> in *.narrowPeak<span class="kw">;</span> <span class="kw">do</span></a>
<a class="sourceLine" id="cb10-3" title="3">  <span class="fu">awk</span> <span class="st">'{if ($5 &gt; 1000) $5 = 1000; print $0}'</span> <span class="st">"</span><span class="va">$f</span><span class="st">"</span> <span class="op">&gt;</span> <span class="st">"</span><span class="va">$f</span><span class="st">"</span>.fix<span class="kw">;</span></a>
<a class="sourceLine" id="cb10-4" title="4">  <span class="ex">bedToBigBed</span> -as=bigNarrowPeak.as -type=bed6+4 <span class="st">"</span><span class="va">$f</span><span class="st">"</span>.fix hg19.chrom.sizes <span class="st">"</span><span class="va">$f</span><span class="st">"</span>.bb<span class="kw">;</span></a>
<a class="sourceLine" id="cb10-5" title="5"><span class="kw">done</span></a></code></pre></div>
</div>
<div id="analysis-with-r" class="section level2">
<h2 class="hasAnchor">
<a href="#analysis-with-r" class="anchor"></a>Analysis with R</h2>
<p>Install this package if you haven’t already with <code><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html">devtools::install_github("j-andrews7/OneLinerOmics")</a></code>. If you don’t have <code>devtools</code>, it can be installed with <code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("devtools")</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="co"># R</span></a>
<a class="sourceLine" id="cb11-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(OneLinerOmics))</a></code></pre></div>
<div id="qc" class="section level3">
<h3 class="hasAnchor">
<a href="#qc" class="anchor"></a>QC</h3>
<p>First, we’ll QC samples, as ChIP tends to be much more variable than RNA-seq in terms of quality. In particular, we’re interested in cross-correlation values and the fraction of reads in peaks, which will help to identify poor quality samples where our IP just didn’t work all that well.</p>
<p>These <em>are</em> difficult samples to work with, after all, and we had very low numbers of cells for many of them. Similar to the RNA-seq analysis, we will use a sample sheet, but this one requires a more strict structure. The columns shown <em>are the only ones allowed</em>. We will also use these sample sheets later on, so if you have more complicated comparisons you want to do, be sure to structure your sheet appropriately so that they are modeled fully within one column.</p>
<p>Pay careful attention to provide the correct file paths to your peak and read files as well as the input controls.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1"><span class="co"># R</span></a>
<a class="sourceLine" id="cb12-2" title="2"><span class="co"># Basic sample sheet.</span></a>
<a class="sourceLine" id="cb12-3" title="3">samples &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="st">"./ChIP_Seq/ChIPQC/H3AC/SampleSheet_H3AC_QC_narrow.NoCTRL.csv"</span>, <span class="dt">header=</span><span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">","</span>)</a>
<a class="sourceLine" id="cb12-4" title="4"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(samples, <span class="dv">10</span>)</a></code></pre></div>
<p>I use <code>ChIPQC</code> to run QC. This package can be a…challenge to use. It is capable of generating a report, but often errors, so this package just uses metrics without the report.</p>
<p>The authors are also slow to keep it up to date, so new R/Bioconductor versions break it all the time. I actually had to download it from <a href="https://github.com/shengqh/ChIPQC">another repo</a> that included a fix to get it to run this time around. Installing this package should install that version, so you shouldn’t have to mess with it.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"><span class="co"># R</span></a>
<a class="sourceLine" id="cb13-2" title="2">h3ac.exp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunChIPQC.html">RunChIPQC</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/ChIPQC/H3AC/"</span>, </a>
<a class="sourceLine" id="cb13-3" title="3">  <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/ChIPQC/H3AC/SampleSheet_H3AC_QC_narrow.NoCTRL.csv"</span>)</a>
<a class="sourceLine" id="cb13-4" title="4"></a>
<a class="sourceLine" id="cb13-5" title="5">faire.exp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunChIPQC.html">RunChIPQC</a></span>(<span class="st">"./ChIP_Seq/ChIPQC/FAIRE/"</span>,</a>
<a class="sourceLine" id="cb13-6" title="6">  <span class="st">"./ChIP_Seq/ChIPQC/FAIRE/SampleSheet_FAIRE_QC_narrow.NoCTRL.csv"</span>)</a>
<a class="sourceLine" id="cb13-7" title="7"></a>
<a class="sourceLine" id="cb13-8" title="8">h3k4me1.exp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunChIPQC.html">RunChIPQC</a></span>(<span class="st">"./ChIP_Seq/ChIPQC/H3K4ME1/"</span>,</a>
<a class="sourceLine" id="cb13-9" title="9">  <span class="st">"./ChIP_Seq/ChIPQC/H3K4ME1/SampleSheet_H3K4ME1_QC_narrow.NoCTRL.csv"</span>)</a>
<a class="sourceLine" id="cb13-10" title="10"></a>
<a class="sourceLine" id="cb13-11" title="11">h3k27ac.exp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunChIPQC.html">RunChIPQC</a></span>(<span class="st">"./ChIP_Seq/ChIPQC/H3K27AC/"</span>,</a>
<a class="sourceLine" id="cb13-12" title="12">  <span class="st">"./ChIP_Seq/ChIPQC/H3K27AC/SampleSheet_H3K27AC_QC_narrow.NoCTRL.csv"</span>)</a></code></pre></div>
<p>In addition to the metrics, this function will generate a plot of the peak overlaps between samples, showing how overlap criteria will impact your consensus peak set. By default, any peaks that overlap in at least 2 samples will be retained and merged. Making this more stringent will help reduce noise at the cost of potentially leaving out real peaks that are unique to a few samples. This curve usually exhibits a near geometric drop-off, with the number of overlapping peaks halving for each increase in strictness. If the drop-off is very steep, it likely means many of your peaks are false positives (or your samples are highly variable). High peak agreement will lead to a less severe drop-off.</p>
</div>
<div id="remove-poor-quality-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#remove-poor-quality-samples" class="anchor"></a>Remove poor quality samples</h3>
<p>Kick ’em off the sample sheets. I removed samples with the following metrics: * FAIRE - &lt;3.5% Reads in Peaks * H3AC - &lt;10% Reads in Peaks * H3K4ME1 - &lt;9.5% Reads in Peaks * H3K27AC - &lt;6.5% Reads in Peaks</p>
<p>Picked rather arbitrarily, but generally want at least 2-3x the typical RiP of the input samples.</p>
</div>
<div id="perform-differential-binding-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-differential-binding-analysis" class="anchor"></a>Perform differential binding analysis</h3>
<p>Another one-liner. This analysis uses <a href="https://bioconductor.org/packages/release/bioc/html/DiffBind.html">DiffBind</a> along with <a href="https://www.bioconductor.org/packages/release/bioc/html/ChIPseeker.html">ChIPseeker</a> for peak annotation and <a href="https://cran.r-project.org/web/packages/enrichR/index.html">enrichR</a> (wrapped by <a href="https://github.com/j-andrews7/EZscRNA">EZscRNA</a>) for enrichment analyses.</p>
<p>I have 4 different marks of interest here, so I’m going to run it on each of them as well as the super enhancers from the H3k27ac data. From our QC up above, I know that I want to bump up the number of samples containing a peak for it to be merged, so I set <code>n.consensus = 3</code> for all but the super enhancers.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="co"># R</span></a>
<a class="sourceLine" id="cb14-2" title="2"><span class="co"># For annotation.</span></a>
<a class="sourceLine" id="cb14-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(TxDb.Hsapiens.UCSC.hg19.knownGene)</a>
<a class="sourceLine" id="cb14-4" title="4">txdb &lt;-<span class="st"> </span>TxDb.Hsapiens.UCSC.hg19.knownGene</a>
<a class="sourceLine" id="cb14-5" title="5"></a>
<a class="sourceLine" id="cb14-6" title="6">ses &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons"</span>,</a>
<a class="sourceLine" id="cb14-7" title="7">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/SampleSheet_H3K27AC_DiffBind_SEs.csv"</span>, </a>
<a class="sourceLine" id="cb14-8" title="8">                   <span class="dt">txdb =</span> txdb, </a>
<a class="sourceLine" id="cb14-9" title="9">                   <span class="dt">level =</span> <span class="st">"Tissue"</span>, </a>
<a class="sourceLine" id="cb14-10" title="10">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb14-11" title="11">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb14-12" title="12">                   <span class="dt">heatmap.preset =</span> <span class="st">"BrTe"</span>,</a>
<a class="sourceLine" id="cb14-13" title="13">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb14-14" title="14"></a>
<a class="sourceLine" id="cb14-15" title="15">h3ac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/H3AC/AllComparisons"</span>,</a>
<a class="sourceLine" id="cb14-16" title="16">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/H3AC/SampleSheet_H3AC_DiffBind.csv"</span>, </a>
<a class="sourceLine" id="cb14-17" title="17">                   <span class="dt">txdb =</span> txdb,</a>
<a class="sourceLine" id="cb14-18" title="18">                   <span class="dt">n.consensus =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb14-19" title="19">                   <span class="dt">se =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt"</span>,</a>
<a class="sourceLine" id="cb14-20" title="20">                   <span class="dt">level =</span> <span class="st">"Tissue"</span>, </a>
<a class="sourceLine" id="cb14-21" title="21">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb14-22" title="22">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb14-23" title="23">                   <span class="dt">heatmap.preset =</span> <span class="st">"OrPu"</span>,</a>
<a class="sourceLine" id="cb14-24" title="24">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb14-25" title="25">faire &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/FAIRE/AllComparisons"</span>,</a>
<a class="sourceLine" id="cb14-26" title="26">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/FAIRE/SampleSheet_FAIRE_DiffBind.csv"</span>, </a>
<a class="sourceLine" id="cb14-27" title="27">                   <span class="dt">txdb =</span> txdb,</a>
<a class="sourceLine" id="cb14-28" title="28">                   <span class="dt">n.consensus =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb14-29" title="29">                   <span class="dt">se =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt"</span>,</a>
<a class="sourceLine" id="cb14-30" title="30">                   <span class="dt">level =</span> <span class="st">"Tissue"</span>, </a>
<a class="sourceLine" id="cb14-31" title="31">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb14-32" title="32">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb14-33" title="33">                   <span class="dt">heatmap.preset =</span> <span class="st">"PuGr"</span>,</a>
<a class="sourceLine" id="cb14-34" title="34">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb14-35" title="35">h3k4me1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K4ME1/AllComparisons"</span>,</a>
<a class="sourceLine" id="cb14-36" title="36">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K4ME1/SampleSheet_H3K4ME1_DiffBind.csv"</span>, </a>
<a class="sourceLine" id="cb14-37" title="37">                   <span class="dt">txdb =</span> txdb,</a>
<a class="sourceLine" id="cb14-38" title="38">                   <span class="dt">n.consensus =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb14-39" title="39">                   <span class="dt">se =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt"</span>,</a>
<a class="sourceLine" id="cb14-40" title="40">                   <span class="dt">level =</span> <span class="st">"Tissue"</span>, </a>
<a class="sourceLine" id="cb14-41" title="41">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb14-42" title="42">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb14-43" title="43">                   <span class="dt">heatmap.preset =</span> <span class="st">"BuOr"</span>,</a>
<a class="sourceLine" id="cb14-44" title="44">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb14-45" title="45">h3k27ac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC/AllComparisons"</span>,</a>
<a class="sourceLine" id="cb14-46" title="46">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC/SampleSheet_H3K27AC_DiffBind.csv"</span>, </a>
<a class="sourceLine" id="cb14-47" title="47">                   <span class="dt">txdb =</span> txdb,</a>
<a class="sourceLine" id="cb14-48" title="48">                   <span class="dt">n.consensus =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb14-49" title="49">                   <span class="dt">se =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt"</span>,</a>
<a class="sourceLine" id="cb14-50" title="50">                   <span class="dt">level =</span> <span class="st">"Tissue"</span>, </a>
<a class="sourceLine" id="cb14-51" title="51">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb14-52" title="52">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb14-53" title="53">                   <span class="dt">heatmap.preset =</span> <span class="st">"BrTe"</span>,</a>
<a class="sourceLine" id="cb14-54" title="54">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<div id="merging-marks" class="section level4">
<h4 class="hasAnchor">
<a href="#merging-marks" class="anchor"></a>Merging Marks</h4>
<p>In this case, I want to generate a peakset to be used across all the marks for differential binding, allowing for easier identification of peaks with multiple marks altered between two groups of samples. Since I’ve already removed a fair amount of noise by setting <code>n.consensus = 3</code>, I’m just going to concatenate and merge the consensus peaks from each mark. Then I’ll edit my samplesheets above to use this new peakset for each sample and re-run the differential binding analysis for each mark.</p>
<p>Here I’m just concatenating and merging the position from a random comparison - it’ll be the same positions across all comparisons, as we’re not filtering at all, so it doesn’t matter which is used.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb15-1" title="1"><span class="co"># bash</span></a>
<a class="sourceLine" id="cb15-2" title="2"><span class="fu">tail</span> -n +2 -q ./ChIP_Seq/DiffBind/H3K27AC/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt \</a>
<a class="sourceLine" id="cb15-3" title="3">  ./ChIP_Seq/DiffBind/FAIRE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt \</a>
<a class="sourceLine" id="cb15-4" title="4">  ./ChIP_Seq/DiffBind/H3AC/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt \</a>
<a class="sourceLine" id="cb15-5" title="5">  ./ChIP_Seq/DiffBind/H3K4ME1/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt <span class="kw"><a href="https://rdrr.io/r/base/Logic.html">|</a></span> <span class="kw">\</span></a>
<a class="sourceLine" id="cb15-6" title="6">  <span class="fu">cut</span> -f 1-3 - <span class="kw"><a href="https://rdrr.io/r/base/Logic.html">|</a></span> <span class="fu">sort</span> -k1,1V -k2,2n - <span class="op">&gt;</span>  ./ChIP_Seq/DiffBind/MergedPeaks/All.Peaks.bed</a>
<a class="sourceLine" id="cb15-7" title="7"></a>
<a class="sourceLine" id="cb15-8" title="8"><span class="ex">bedtools</span> merge ./ChIP_Seq/DiffBind/MergedPeaks/All.Peaks.bed <span class="op">&gt;</span> ./ChIP_Seq/DiffBind/MergedPeaks/MergedPeaks.bed</a></code></pre></div>
<p>Then I edited my sample sheets to use the merged peakset for all samples.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1"><span class="co"># R</span></a>
<a class="sourceLine" id="cb16-2" title="2"><span class="co"># Basic sample sheet.</span></a>
<a class="sourceLine" id="cb16-3" title="3">samples &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3AC/SampleSheet_H3AC_DiffBind.csv"</span>, <span class="dt">header=</span><span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">","</span>)</a>
<a class="sourceLine" id="cb16-4" title="4"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(samples, <span class="dv">10</span>)</a></code></pre></div>
</div>
</div>
<div id="running-diffbind-on-a-merged-consensus-peakset" class="section level3">
<h3 class="hasAnchor">
<a href="#running-diffbind-on-a-merged-consensus-peakset" class="anchor"></a>Running DiffBind on a Merged, Consensus Peakset</h3>
<p>Now the analysis can be run again using the new samplesheets to generate results for the same peaks across all marks.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1"><span class="co"># R</span></a>
<a class="sourceLine" id="cb17-2" title="2">ses &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons"</span>,</a>
<a class="sourceLine" id="cb17-3" title="3">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/SampleSheet_H3K27AC_DiffBind_SEs.csv"</span>, </a>
<a class="sourceLine" id="cb17-4" title="4">                   <span class="dt">txdb =</span> txdb, </a>
<a class="sourceLine" id="cb17-5" title="5">                   <span class="dt">dba =</span> ses,</a>
<a class="sourceLine" id="cb17-6" title="6">                   <span class="dt">level =</span> <span class="st">"Tissue"</span>, </a>
<a class="sourceLine" id="cb17-7" title="7">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb17-8" title="8">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb17-9" title="9">                   <span class="dt">heatmap.preset =</span> <span class="st">"BrTe"</span>,</a>
<a class="sourceLine" id="cb17-10" title="10">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb17-11" title="11"></a>
<a class="sourceLine" id="cb17-12" title="12">h3ac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3AC/AllComparisons"</span>,</a>
<a class="sourceLine" id="cb17-13" title="13">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3AC/SampleSheet_H3AC_DiffBind.csv"</span>, </a>
<a class="sourceLine" id="cb17-14" title="14">                   <span class="dt">txdb =</span> txdb,</a>
<a class="sourceLine" id="cb17-15" title="15">                   <span class="dt">dba =</span> h3ac,</a>
<a class="sourceLine" id="cb17-16" title="16">                   <span class="dt">n.consensus =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb17-17" title="17">                   <span class="dt">se =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt"</span>,</a>
<a class="sourceLine" id="cb17-18" title="18">                   <span class="dt">level =</span> <span class="st">"Tissue"</span>, </a>
<a class="sourceLine" id="cb17-19" title="19">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb17-20" title="20">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb17-21" title="21">                   <span class="dt">heatmap.preset =</span> <span class="st">"OrPu"</span>,</a>
<a class="sourceLine" id="cb17-22" title="22">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb17-23" title="23">faire &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/FAIRE/AllComparisons"</span>,</a>
<a class="sourceLine" id="cb17-24" title="24">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/FAIRE/SampleSheet_FAIRE_DiffBind.csv"</span>, </a>
<a class="sourceLine" id="cb17-25" title="25">                   <span class="dt">txdb =</span> txdb,</a>
<a class="sourceLine" id="cb17-26" title="26">                   <span class="dt">dba =</span> faire,</a>
<a class="sourceLine" id="cb17-27" title="27">                   <span class="dt">n.consensus =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb17-28" title="28">                   <span class="dt">se =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt"</span>,</a>
<a class="sourceLine" id="cb17-29" title="29">                   <span class="dt">level =</span> <span class="st">"Tissue"</span>, </a>
<a class="sourceLine" id="cb17-30" title="30">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb17-31" title="31">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb17-32" title="32">                   <span class="dt">heatmap.preset =</span> <span class="st">"PuGr"</span>,</a>
<a class="sourceLine" id="cb17-33" title="33">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb17-34" title="34">h3k4me1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3K4ME1/AllComparisons"</span>,</a>
<a class="sourceLine" id="cb17-35" title="35">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3K4ME1/SampleSheet_H3K4ME1_DiffBind.csv"</span>, </a>
<a class="sourceLine" id="cb17-36" title="36">                   <span class="dt">txdb =</span> txdb,</a>
<a class="sourceLine" id="cb17-37" title="37">                   <span class="dt">dba =</span> h3k4me1,</a>
<a class="sourceLine" id="cb17-38" title="38">                   <span class="dt">n.consensus =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb17-39" title="39">                   <span class="dt">se =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt"</span>,</a>
<a class="sourceLine" id="cb17-40" title="40">                   <span class="dt">level =</span> <span class="st">"Tissue"</span>, </a>
<a class="sourceLine" id="cb17-41" title="41">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb17-42" title="42">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb17-43" title="43">                   <span class="dt">heatmap.preset =</span> <span class="st">"BuOr"</span>,</a>
<a class="sourceLine" id="cb17-44" title="44">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb17-45" title="45">h3k27ac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3K27AC/AllComparisons"</span>,</a>
<a class="sourceLine" id="cb17-46" title="46">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3K27AC/SampleSheet_H3K27AC_DiffBind.csv"</span>, </a>
<a class="sourceLine" id="cb17-47" title="47">                   <span class="dt">txdb =</span> txdb,</a>
<a class="sourceLine" id="cb17-48" title="48">                   <span class="dt">dba =</span> h3k27ac,</a>
<a class="sourceLine" id="cb17-49" title="49">                   <span class="dt">n.consensus =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb17-50" title="50">                   <span class="dt">se =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt"</span>,</a>
<a class="sourceLine" id="cb17-51" title="51">                   <span class="dt">level =</span> <span class="st">"Tissue"</span>, </a>
<a class="sourceLine" id="cb17-52" title="52">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb17-53" title="53">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb17-54" title="54">                   <span class="dt">heatmap.preset =</span> <span class="st">"BrTe"</span>,</a>
<a class="sourceLine" id="cb17-55" title="55">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p>I also want to run another comparison (CANCER v NORMAL), so I will set that up as well.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1"><span class="co"># R</span></a>
<a class="sourceLine" id="cb18-2" title="2">ses &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/CANCERvNORMAL"</span>,</a>
<a class="sourceLine" id="cb18-3" title="3">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/SampleSheet_H3K27AC_DiffBind_SEs.csv"</span>, </a>
<a class="sourceLine" id="cb18-4" title="4">                   <span class="dt">txdb =</span> txdb, </a>
<a class="sourceLine" id="cb18-5" title="5">                   <span class="dt">dba =</span> ses,</a>
<a class="sourceLine" id="cb18-6" title="6">                   <span class="dt">level =</span> <span class="st">"Condition"</span>, </a>
<a class="sourceLine" id="cb18-7" title="7">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb18-8" title="8">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.5</span>, <span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb18-9" title="9">                   <span class="dt">heatmap.preset =</span> <span class="st">"BrTe"</span>,</a>
<a class="sourceLine" id="cb18-10" title="10">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb18-11" title="11"></a>
<a class="sourceLine" id="cb18-12" title="12">h3ac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3AC/CANCERvNORMAL"</span>,</a>
<a class="sourceLine" id="cb18-13" title="13">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3AC/SampleSheet_H3AC_DiffBind.csv"</span>, </a>
<a class="sourceLine" id="cb18-14" title="14">                   <span class="dt">txdb =</span> txdb,</a>
<a class="sourceLine" id="cb18-15" title="15">                   <span class="dt">dba =</span> h3ac,</a>
<a class="sourceLine" id="cb18-16" title="16">                   <span class="dt">n.consensus =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb18-17" title="17">                   <span class="dt">se =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt"</span>,</a>
<a class="sourceLine" id="cb18-18" title="18">                   <span class="dt">level =</span> <span class="st">"Condition"</span>, </a>
<a class="sourceLine" id="cb18-19" title="19">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb18-20" title="20">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb18-21" title="21">                   <span class="dt">heatmap.preset =</span> <span class="st">"OrPu"</span>,</a>
<a class="sourceLine" id="cb18-22" title="22">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb18-23" title="23">faire &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/FAIRE/CANCERvNORMAL"</span>,</a>
<a class="sourceLine" id="cb18-24" title="24">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/FAIRE/SampleSheet_FAIRE_DiffBind.csv"</span>, </a>
<a class="sourceLine" id="cb18-25" title="25">                   <span class="dt">txdb =</span> txdb,</a>
<a class="sourceLine" id="cb18-26" title="26">                   <span class="dt">dba =</span> faire,</a>
<a class="sourceLine" id="cb18-27" title="27">                   <span class="dt">n.consensus =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb18-28" title="28">                   <span class="dt">se =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt"</span>,</a>
<a class="sourceLine" id="cb18-29" title="29">                   <span class="dt">level =</span> <span class="st">"Condition"</span>, </a>
<a class="sourceLine" id="cb18-30" title="30">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb18-31" title="31">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb18-32" title="32">                   <span class="dt">heatmap.preset =</span> <span class="st">"PuGr"</span>,</a>
<a class="sourceLine" id="cb18-33" title="33">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb18-34" title="34">h3k4me1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3K4ME1/CANCERvNORMAL"</span>,</a>
<a class="sourceLine" id="cb18-35" title="35">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3K4ME1/SampleSheet_H3K4ME1_DiffBind.csv"</span>, </a>
<a class="sourceLine" id="cb18-36" title="36">                   <span class="dt">txdb =</span> txdb,</a>
<a class="sourceLine" id="cb18-37" title="37">                   <span class="dt">dba =</span> h3k4me1,</a>
<a class="sourceLine" id="cb18-38" title="38">                   <span class="dt">n.consensus =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb18-39" title="39">                   <span class="dt">se =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt"</span>,</a>
<a class="sourceLine" id="cb18-40" title="40">                   <span class="dt">level =</span> <span class="st">"Condition"</span>, </a>
<a class="sourceLine" id="cb18-41" title="41">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb18-42" title="42">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb18-43" title="43">                   <span class="dt">heatmap.preset =</span> <span class="st">"BuOr"</span>,</a>
<a class="sourceLine" id="cb18-44" title="44">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb18-45" title="45">h3k27ac &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDiffBind.html">RunDiffBind</a></span>(<span class="dt">outpath =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3K27AC/CANCERvNORMAL"</span>,</a>
<a class="sourceLine" id="cb18-46" title="46">                   <span class="dt">samplesheet =</span> <span class="st">"./ChIP_Seq/DiffBind/MergedPeaks/H3K27AC/SampleSheet_H3K27AC_DiffBind.csv"</span>, </a>
<a class="sourceLine" id="cb18-47" title="47">                   <span class="dt">txdb =</span> txdb,</a>
<a class="sourceLine" id="cb18-48" title="48">                   <span class="dt">dba =</span> h3k27ac,</a>
<a class="sourceLine" id="cb18-49" title="49">                   <span class="dt">n.consensus =</span> <span class="dv">1</span>,</a>
<a class="sourceLine" id="cb18-50" title="50">                   <span class="dt">se =</span> <span class="st">"./ChIP_Seq/DiffBind/H3K27AC_SE/AllComparisons/NoBlock/ResultsTables/CLL-v-DL.AllPeaks.txt"</span>,</a>
<a class="sourceLine" id="cb18-51" title="51">                   <span class="dt">level =</span> <span class="st">"Condition"</span>, </a>
<a class="sourceLine" id="cb18-52" title="52">                   <span class="dt">fdr.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>),</a>
<a class="sourceLine" id="cb18-53" title="53">                   <span class="dt">fc.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb18-54" title="54">                   <span class="dt">heatmap.preset =</span> <span class="st">"BrTe"</span>,</a>
<a class="sourceLine" id="cb18-55" title="55">                   <span class="dt">scale.full =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
</div>
</div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session Information</h1>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</a></code></pre></div>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 10 x64 (build 16299)
## 
## Matrix products: default
## 
## locale:
## [1] LC_COLLATE=English_United States.1252 
## [2] LC_CTYPE=English_United States.1252   
## [3] LC_MONETARY=English_United States.1252
## [4] LC_NUMERIC=C                          
## [5] LC_TIME=English_United States.1252    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] BiocStyle_2.14.0
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.2         rstudioapi_0.10    knitr_1.25        
##  [4] magrittr_1.5       MASS_7.3-51.4      lattice_0.20-38   
##  [7] R6_2.4.0           rlang_0.4.1        stringr_1.4.0     
## [10] tools_3.6.1        grid_3.6.1         xfun_0.10         
## [13] htmltools_0.4.0    yaml_2.2.0         assertthat_0.2.1  
## [16] digest_0.6.22      rprojroot_1.3-2    pkgdown_1.4.1     
## [19] crayon_1.3.4       bookdown_0.14      Matrix_1.2-17     
## [22] BiocManager_1.30.9 fs_1.3.1           memoise_1.1.0     
## [25] evaluate_0.14      rmarkdown_1.16     stringi_1.4.3     
## [28] compiler_3.6.1     desc_1.2.0         backports_1.1.5   
## [31] jsonlite_1.6       reticulate_1.13</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#faqsuggested-reading">FAQ/Suggested Reading</a><ul class="nav nav-pills nav-stacked">
<li><a href="#alignment-indexing-and-blacklisted-read-removal">Alignment, Indexing, and Blacklisted Read Removal</a></li>
      <li><a href="#peak-calling">Peak Calling</a></li>
      <li><a href="#super-enhancer-calling-with-rose">Super Enhancer Calling with ROSE</a></li>
      <li><a href="#making-tracks">Making Tracks</a></li>
      <li><a href="#analysis-with-r">Analysis with R</a></li>
      </ul>
</li>
      <li><a href="#session-information">Session Information</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://j-andrews7.github.io">Jared Andrews</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
