<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyzing RNA-seq data • AndrewsBCellLymphoma</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing RNA-seq data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">AndrewsBCellLymphoma</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/RNAseqAnalysis.html">Analyzing RNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/j-andrews7/AndrewsBCellLymphoma">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Analyzing RNA-seq data</h1>
                        <h4 class="author">Jared Andrews</h4>
            <address class="author_afil">
      Washington University School of Medicine, Department of Laboratory and Genomic Medicine<br><a class="author_email" href="mailto:#"></a><a href="mailto:jared.andrews07@gmail.com" class="email">jared.andrews07@gmail.com</a>
      </address>
                  
            <h4 class="date">Revised: Ocotober 22nd, 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/j-andrews7/AndrewsBCellLymphoma/blob/master/vignettes/RNAseqAnalysis.Rmd"><code>vignettes/RNAseqAnalysis.Rmd</code></a></small>
      <div class="hidden name"><code>RNAseqAnalysis.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This vignette walks through analysis of the B cell lymphoma RNA-seq data included in the future Andrews et al. (2019, your favorite journal with an impact factor &gt; 30) paper. It uses the convenience functions included in this <a href="https://github.com/j-andrews7/AndrewsBCellLymphoma">package</a> to streamline the process.</p>
<p>Several command line tools as well as python are also utilized - each code block should indicate whether it is <code>bash</code>, <code>R</code>, or <code>python</code> code. If your system is set up properly, you can still <a href="https://bookdown.org/yihui/rmarkdown/language-engines.html">run these code chunks in Rmarkdown files within RStudio</a>, assuming you have all of the software installed.</p>
<p>Required Software:</p>
<ul>
<li>R (3.6+)</li>
<li>python</li>
<li><a href="https://combine-lab.github.io/salmon/">salmon (v0.14+)</a></li>
<li>
<a href="https://bedtools.readthedocs.io/en/latest/">bedtools</a> - only needed if generating salmon decoy transcriptome</li>
<li>
<a href="https://github.com/marbl/MashMap">mashmap</a> - only needed if generating salmon decoy transcriptome</li>
</ul>
</div>
<div id="suggested-reading" class="section level1">
<h1 class="hasAnchor">
<a href="#suggested-reading" class="anchor"></a>Suggested Reading</h1>
<p>That is, if you ask me a question and haven’t read these first, I’m going to tell you to read them.</p>
<ul>
<li>The <a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#standard-workflow">DESeq2 vignette</a>
</li>
<li>The <a href="http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html">rnaseqGene</a> workflow, which walks through a DESeq2 analysis in great detail.</li>
</ul>
<div id="generating-gene-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#generating-gene-counts" class="anchor"></a>Generating gene counts</h2>
<p>Going the <code>salmon</code> -&gt; <code>DESeq2</code> route. This requires R, salmon v0.14, and DESeq2 (along with a few other packages). Using hg19, cause forget re-aligning all the ChIP-seq and making new tracks. <code>salmon</code> needs <code>mashmap</code> and <code>bedtools</code> as well if generating “decoy-aware” transcriptomes. The Google can show you how to install those.</p>
<div id="create-decoy-reference-transcriptome" class="section level3">
<h3 class="hasAnchor">
<a href="#create-decoy-reference-transcriptome" class="anchor"></a>Create decoy reference transcriptome</h3>
<p>Newer versions of salmon recommend using a <em>decoy-aware</em> transcriptome to generate the index file to be used for selective alignment. This process is a bit annoying, so the authors have pre-computed some <a href="https://drive.google.com/drive/folders/14VqSdZAKH82QwDWhMXNLFqMoskoqv3fS">decoy transcriptomes</a> for hg19 and hg38 with different annotations. In this case, I used the <a href="ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/GRCh37_mapping/">GENCODE v31 annotations</a> that had been lifted over to hg19, so I had to generate them myself.</p>
<blockquote>
<p><strong>This had to be run on a cluster with 128 GB RAM. If you can avoid this step by using their pre-generated decoys, I recommend doing so.</strong></p>
</blockquote>
<p>My data is already in folders and all, so some of the paths are kinda inefficient, but I really don’t care. This is needed to create the index for <code>salmon</code>. Using GENCODE v31 annotations/sequences that have been mapped back to GRCh37 (hg19).</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># bash</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">mkdir</span> -p RNA_seq/ref</span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="co"># annotations</span></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">wget</span> ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/GRCh37_mapping/gencode.v31lift37.annotation.gtf.gz</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="fu">mv</span> gencode.v31lift37.annotation.gtf.gz ./RNA_Seq/ref/gencode.v31lift37.annotation.gtf.gz</span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="fu">gunzip</span> ./RNA_seq/ref/gencode.v31lift37.annotation.gtf.gz</span>
<span id="cb1-8"><a href="#cb1-8"></a></span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="co"># transcriptome</span></span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="fu">wget</span> ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/GRCh37_mapping/gencode.v31lift37.transcripts.fa.gz</span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="fu">mv</span> gencode.v31lift37.transcripts.fa.gz ./RNA_Seq/ref/gencode.v31lift37.transcripts.fa.gz</span>
<span id="cb1-12"><a href="#cb1-12"></a><span class="fu">gunzip</span> ./RNA_seq/ref/gencode.v31lift37.transcripts.fa.gz</span>
<span id="cb1-13"><a href="#cb1-13"></a></span>
<span id="cb1-14"><a href="#cb1-14"></a><span class="co"># genome</span></span>
<span id="cb1-15"><a href="#cb1-15"></a><span class="fu">wget</span> ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/GRCh37_mapping/GRCh37.primary_assembly.genome.fa.gz</span>
<span id="cb1-16"><a href="#cb1-16"></a><span class="fu">mv</span> GRCh37.primary_assembly.genome.fa.gz ./RNA_Seq/ref/GRCh37.primary_assembly.genome.fa.gz</span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="fu">gunzip</span> ./RNA_seq/ref/GRCh37.primary_assembly.genome.fa.gz</span></code></pre></div>
<p>Download the script to generate the decoys.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># bash</span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="co"># Script for generating decoy genome.</span></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="fu">wget</span> https://raw.githubusercontent.com/COMBINE-lab/SalmonTools/master/scripts/generateDecoyTranscriptome.sh</span></code></pre></div>
<p>This script crashed a lot for me, so I ended up editing it to just snag the last 10 commands or so (those with the step comments) and removing the variable parsing at the beginning, as it uses some deprecated linux commands (<code>readpath</code>) that I couldn’t install on our computing cluster.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># bash</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="fu">bash</span> generateDecoyTranscriptome.sh -j 2 -a ./RNA_Seq/ref/gencode.v31lift37.annotation.gtf -g ./RNA_Seq/ref/GRCh37.primary_assembly.genome.fa -t ./RNA_Seq/ref/gencode.v31lift37.transcripts.fa -o ./RNA_Seq/ref/decoys/</span></code></pre></div>
</div>
<div id="index-the-transcriptome" class="section level3">
<h3 class="hasAnchor">
<a href="#index-the-transcriptome" class="anchor"></a>Index the transcriptome</h3>
<p>This will generate the <em>decoy-aware</em> transcriptome index for our annotations.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># Bash</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co"># Now actually create the index for salmon. We lower k as some of our samples use 50 bp reads.</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="ex">salmon</span> index -t ./RNA_Seq/ref/decoys/gentrome.fa -i ./RNA_Seq/ref/gencodev31_grch37_index --decoys ./RNA_Seq/ref/decoys/decoys.txt -k 25 --gencode</span></code></pre></div>
</div>
<div id="quantify" class="section level3">
<h3 class="hasAnchor">
<a href="#quantify" class="anchor"></a>Quantify</h3>
<p>Actually generate the gene counts now.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># bash</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="kw"><a href="https://rdrr.io/r/base/Control.html">for</a></span> <span class="ex">f</span> in ./RNA_Seq/FASTQs/*<span class="kw">;</span> <span class="kw">do</span></span>
<span id="cb5-3"><a href="#cb5-3"></a>    <span class="va">samp=</span><span class="st">"</span><span class="va">${f##</span>*/<span class="va">}</span><span class="st">"</span></span>
<span id="cb5-4"><a href="#cb5-4"></a>    <span class="va">count=$(</span><span class="fu">ls</span> -1 <span class="st">"</span><span class="va">$f</span><span class="st">"</span> <span class="kw"><a href="https://rdrr.io/r/base/Logic.html">|</a></span> <span class="fu">wc</span> -l<span class="va">)</span></span>
<span id="cb5-5"><a href="#cb5-5"></a>    <span class="bu">echo</span> <span class="st">"</span><span class="va">$count</span><span class="st">"</span></span>
<span id="cb5-6"><a href="#cb5-6"></a>    <span class="kw"><a href="https://rdrr.io/r/base/Control.html">if</a></span><span class="bu"> [</span> <span class="va">$count</span> <span class="ot">-gt</span> 1<span class="bu"> ]</span></span>
<span id="cb5-7"><a href="#cb5-7"></a>    <span class="kw">then</span></span>
<span id="cb5-8"><a href="#cb5-8"></a>        <span class="ex">salmon</span> quant --validateMappings --threads 2 --seqBias -l A -1 <span class="st">"</span><span class="va">$f</span><span class="st">"</span>/<span class="st">"</span><span class="va">$samp</span><span class="st">"</span>_1.fq.gz -2 <span class="st">"</span><span class="va">$f</span><span class="st">"</span>/<span class="st">"</span><span class="va">$samp</span><span class="st">"</span>_2.fq.gz -o ./RNA_Seq/quants/<span class="st">"</span><span class="va">$samp</span><span class="st">"</span> --index ./RNA_Seq/ref/gencodev31_grch37_index</span>
<span id="cb5-9"><a href="#cb5-9"></a>    <span class="kw"><a href="https://rdrr.io/r/base/Control.html">else</a></span></span>
<span id="cb5-10"><a href="#cb5-10"></a>        <span class="ex">salmon</span> quant --validateMappings --threads 2 --seqBias -l A -r <span class="st">"</span><span class="va">$f</span><span class="st">"</span>/<span class="st">"</span><span class="va">$samp</span><span class="st">"</span>.RNA.fq.gz -o ./RNA_Seq/quants/<span class="st">"</span><span class="va">$samp</span><span class="st">"</span> --index ./RNA_Seq/ref/gencodev31_grch37_index</span>
<span id="cb5-11"><a href="#cb5-11"></a>    <span class="kw">fi</span></span>
<span id="cb5-12"><a href="#cb5-12"></a><span class="kw">done</span></span></code></pre></div>
</div>
<div id="make-transcript-to-gene-mappings" class="section level3">
<h3 class="hasAnchor">
<a href="#make-transcript-to-gene-mappings" class="anchor"></a>Make Transcript-to-Gene Mappings</h3>
<p>For gene-level summarization and readability.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1"></a><span class="co"># python</span></span>
<span id="cb6-2"><a href="#cb6-2"></a>outter <span class="op">=</span> <span class="bu">open</span>(<span class="st">"./RNA_Seq/ref/tx2gene.txt"</span>, <span class="st">"w"</span>)</span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">"./RNA_Seq/ref/gencode.v31lift37.transcripts.fa"</span>) <span class="im">as</span> f:</span>
<span id="cb6-4"><a href="#cb6-4"></a>    <span class="cf">for</span> line <span class="kw"><a href="https://rdrr.io/r/base/Control.html">in</a></span> f:</span>
<span id="cb6-5"><a href="#cb6-5"></a>        <span class="cf">if</span> <span class="kw">not</span> line.startswith(<span class="st">"&gt;"</span>):</span>
<span id="cb6-6"><a href="#cb6-6"></a>            <span class="cf">continue</span></span>
<span id="cb6-7"><a href="#cb6-7"></a>        line <span class="op">=</span> line.strip().strip(<span class="st">"&gt;"</span>).split(<span class="st">"|"</span>)</span>
<span id="cb6-8"><a href="#cb6-8"></a>        tx <span class="op">=</span> line[<span class="dv">0</span>]</span>
<span id="cb6-9"><a href="#cb6-9"></a>        symb <span class="op">=</span> line[<span class="dv">5</span>]</span>
<span id="cb6-10"><a href="#cb6-10"></a>        out <span class="op">=</span> <span class="st">"</span><span class="ch">\t</span><span class="st">"</span>.join([tx, symb])</span>
<span id="cb6-11"><a href="#cb6-11"></a>        <span class="bu">print</span>(out, <span class="bu">file</span> <span class="op">=</span> outter)</span>
<span id="cb6-12"><a href="#cb6-12"></a></span>
<span id="cb6-13"><a href="#cb6-13"></a>outter.close()</span></code></pre></div>
</div>
</div>
<div id="analysis-with-r" class="section level2">
<h2 class="hasAnchor">
<a href="#analysis-with-r" class="anchor"></a>Analysis with R</h2>
<p>Install this package if you haven’t already with <code><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html">devtools::install_github("j-andrews7/AndrewsBCellLymphoma")</a></code>. If you don’t have <code>devtools</code>, it can be installed with <code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("devtools")</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># R</span></span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(AndrewsBCellLymphoma))</span></code></pre></div>
<p>The only thing you need now is a sample sheet, which is just a table with sample names and metadata columns. We can take a peek at the sample sheet I’m using so he can see how they’re constructed. It can include as many metadata columns as you’d like.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># R</span></span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="co"># Basic sample sheet.</span></span>
<span id="cb8-3"><a href="#cb8-3"></a>samples &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="st">"./RNA_Seq/Samples.allCLL.txt"</span>, <span class="dt">header=</span><span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(samples, <span class="dv">10</span>)</span></code></pre></div>
<div id="run-deseq2" class="section level3">
<h3 class="hasAnchor">
<a href="#run-deseq2" class="anchor"></a>Run DESeq2</h3>
<p>DESeq2 is a very popular R package for performing differential expression analysis. Raw RNA-seq counts follow a negative binomial distribution, which is how DESeq2 models the data. It is quick, flexible, and generally works quite well with little to no parameter tweaking. It can also help to mitigate the technical or batch effects on the analysis, a feature which we’ll take advantage of here.</p>
<p>Most of these samples are 150 bp, paired-end sequencing (as shown in the ‘lib’ column of the sample sheet). A handful are 50 bp, single-end sequencing, which introduces a significant confounding factor to the analysis. In our model design, we’ll specify that we want to ignore effects due to this technical difference (by setting <code>block = "lib"</code>). The model isn’t magic, so we may very well still see evidence of the library-prep differences regardless, as they are quite strong.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co"># R</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>mydata &lt;-<span class="st"> </span><span class="kw"><a href="../reference/RunDESeq2.html">RunDESeq2</a></span>(<span class="dt">outpath =</span> <span class="st">"./RNA_Seq/Final_Analyses/allCLL"</span>, <span class="dt">quants.path =</span> <span class="st">"./RNA_Seq/quants"</span>, </span>
<span id="cb9-3"><a href="#cb9-3"></a>          <span class="dt">samplesheet =</span> <span class="st">"./RNA_Seq/Samples.allCLL.txt"</span>, <span class="dt">tx2gene =</span> <span class="st">"./RNA_Seq/ref/tx2gene.txt"</span>,</span>
<span id="cb9-4"><a href="#cb9-4"></a>          <span class="dt">level =</span> <span class="st">"disease"</span>, <span class="dt">block =</span> <span class="st">"lib"</span>, <span class="dt">padj.thresh =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.05</span>, <span class="fl">0.01</span>, <span class="fl">0.001</span>), </span>
<span id="cb9-5"><a href="#cb9-5"></a>          <span class="dt">plot.annos =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"disease"</span>, <span class="st">"disease.sub"</span>), <span class="dt">count.filt =</span> <span class="dv">50</span>)</span></code></pre></div>
</div>
<div id="next-steps" class="section level3">
<h3 class="hasAnchor">
<a href="#next-steps" class="anchor"></a>Next Steps</h3>
<p>Yes, that’s it. Explore the plots, tables, enrichments, etc. Re-run if necessary, tweaking parameters as you see fit.</p>
</div>
</div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session Information</h1>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</span></code></pre></div>
<pre><code>## R version 3.6.0 (2019-04-26)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.2 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1
## 
## locale:
##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] AndrewsBCellLymphoma_0.0.0.9000 BiocStyle_2.13.2               
## 
## loaded via a namespace (and not attached):
##   [1] reticulate_1.13             R.utils_2.9.0              
##   [3] tidyselect_0.2.5            RSQLite_2.1.2              
##   [5] AnnotationDbi_1.47.1        htmlwidgets_1.5.1          
##   [7] grid_3.6.0                  BiocParallel_1.19.5        
##   [9] Rtsne_0.15                  munsell_0.5.0              
##  [11] codetools_0.2-16            ica_1.0-2                  
##  [13] preprocessCore_1.47.1       future_1.14.0              
##  [15] colorspace_1.4-1            Biobase_2.45.1             
##  [17] knitr_1.25                  rstudioapi_0.10            
##  [19] Seurat_3.1.1                stats4_3.6.0               
##  [21] SingleCellExperiment_1.7.11 ROCR_1.0-7                 
##  [23] gbRd_0.4-11                 listenv_0.7.0              
##  [25] Rdpack_0.11-0               tximport_1.13.17           
##  [27] GenomeInfoDbData_1.2.1      bit64_0.9-7                
##  [29] pheatmap_1.0.12             rprojroot_1.3-2            
##  [31] vctrs_0.2.0                 xfun_0.10                  
##  [33] R6_2.4.0                    GenomeInfoDb_1.21.2        
##  [35] ggbeeswarm_0.6.0            rsvd_1.0.2                 
##  [37] locfit_1.5-9.1              bitops_1.0-6               
##  [39] DelayedArray_0.11.8         assertthat_0.2.1           
##  [41] SDMTools_1.1-221.1          scales_1.0.0               
##  [43] nnet_7.3-12                 beeswarm_0.2.3             
##  [45] gtable_0.3.0                npsurv_0.4-0               
##  [47] affy_1.63.1                 globals_0.12.4             
##  [49] rlang_0.4.0                 zeallot_0.1.0              
##  [51] genefilter_1.67.1           splines_3.6.0              
##  [53] lazyeval_0.2.2              acepack_1.4.1              
##  [55] checkmate_1.9.4             BiocManager_1.30.9         
##  [57] yaml_2.2.0                  reshape2_1.4.3             
##  [59] backports_1.1.5             Hmisc_4.2-0                
##  [61] tools_3.6.0                 bookdown_0.14              
##  [63] ggplot2_3.2.1               affyio_1.55.0              
##  [65] gplots_3.0.1.1              RColorBrewer_1.1-2         
##  [67] BiocGenerics_0.31.6         ggridges_0.5.1             
##  [69] Rcpp_1.0.2                  plyr_1.8.4                 
##  [71] base64enc_0.1-3             zlibbioc_1.31.0            
##  [73] purrr_0.3.3                 RCurl_1.95-4.12            
##  [75] rpart_4.1-15                pbapply_1.4-2              
##  [77] viridis_0.5.1               cowplot_1.0.0              
##  [79] S4Vectors_0.23.25           zoo_1.8-6                  
##  [81] SummarizedExperiment_1.15.9 ggrepel_0.8.1              
##  [83] cluster_2.0.9               fs_1.3.1                   
##  [85] magrittr_1.5                data.table_1.12.6          
##  [87] lmtest_0.9-37               RANN_2.6.1                 
##  [89] fitdistrplus_1.0-14         matrixStats_0.55.0         
##  [91] lsei_1.2-0                  evaluate_0.14              
##  [93] xtable_1.8-4                XML_3.98-1.20              
##  [95] IRanges_2.19.17             gridExtra_2.3              
##  [97] compiler_3.6.0              scater_1.13.27             
##  [99] tibble_2.1.3                KernSmooth_2.23-15         
## [101] crayon_1.3.4                R.oo_1.22.0                
## [103] htmltools_0.4.0             Formula_1.2-3              
## [105] tidyr_1.0.0                 geneplotter_1.63.0         
## [107] RcppParallel_4.4.4          DBI_1.0.0                  
## [109] MASS_7.3-51.1               Matrix_1.2-17              
## [111] vsn_3.53.0                  R.methodsS3_1.7.1          
## [113] gdata_2.18.0                parallel_3.6.0             
## [115] metap_1.1                   igraph_1.2.4.1             
## [117] GenomicRanges_1.37.17       pkgconfig_2.0.3            
## [119] pkgdown_1.4.1               foreign_0.8-70             
## [121] plotly_4.9.0                annotate_1.63.0            
## [123] vipor_0.4.5                 EnhancedVolcano_1.3.5      
## [125] XVector_0.25.0              bibtex_0.4.2               
## [127] stringr_1.4.0               digest_0.6.22              
## [129] sctransform_0.2.0           RcppAnnoy_0.0.13           
## [131] tsne_0.1-3                  rmarkdown_1.16             
## [133] leiden_0.3.1                enrichR_2.1                
## [135] htmlTable_1.13.2            uwot_0.1.4                 
## [137] DelayedMatrixStats_1.7.2    gtools_3.8.1               
## [139] rjson_0.2.20                lifecycle_0.1.0            
## [141] EZscRNA_0.1.0.9000          nlme_3.1-140               
## [143] jsonlite_1.6                BiocNeighbors_1.3.5        
## [145] limma_3.41.18               desc_1.2.0                 
## [147] viridisLite_0.3.0           pillar_1.4.2               
## [149] lattice_0.20-38             httr_1.4.1                 
## [151] survival_2.43-3             glue_1.3.1                 
## [153] png_0.1-7                   bit_1.1-14                 
## [155] stringi_1.4.3               blob_1.2.0                 
## [157] DESeq2_1.25.17              BiocSingular_1.1.7         
## [159] latticeExtra_0.6-28         caTools_1.17.1.2           
## [161] memoise_1.1.0               dplyr_0.8.3                
## [163] irlba_2.3.3                 future.apply_1.3.0         
## [165] ape_5.3</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#suggested-reading">Suggested Reading</a><ul class="nav nav-pills nav-stacked">
<li><a href="#generating-gene-counts">Generating gene counts</a></li>
      <li><a href="#analysis-with-r">Analysis with R</a></li>
      </ul>
</li>
      <li><a href="#session-information">Session Information</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://j-andrews7.github.io">Jared Andrews</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
