<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ OneLinerOmics</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">OneLinerOmics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/ChIPseqAnalysis.html">Analyzing ChIP-seq data</a>
    </li>
    <li>
      <a href="../articles/RNAseqAnalysis.html">Analyzing RNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/j-andrews7/OneLinerOmics">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-rna-seq-analysis" class="hasAnchor"><a href="#section-rna-seq-analysis" class="anchor"></a>RNA-seq analysis</h2>
          <p class="section-desc"><p>Functions that perform DESeq2 analyses, utilizing the utility and plotting functions to generate output tables and figures. These are the crux of the RNA-seq portion of the package and what most users should use.</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="RunDESeq2.html">RunDESeq2()</a></code> </p>
        </td>
        <td><p>Run DESeq2 analyses</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ProcessDEGs.html">ProcessDEGs()</a></code> </p>
        </td>
        <td><p>Extract, visualize, and save DEG lists</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-rna-seq-plotting" class="hasAnchor"><a href="#section-rna-seq-plotting" class="anchor"></a>RNA-seq Plotting</h2>
          <p class="section-desc"><p>Functions that generate plots for RNA-seq data. Calling these individually can sometimes be useful.</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="PlotRNABoxplots.html">PlotRNABoxplots()</a></code> </p>
        </td>
        <td><p>Plot gene counts as box plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PlotRNACombinedHeatmaps.html">PlotRNACombinedHeatmaps()</a></code> </p>
        </td>
        <td><p>Generate gene heatmaps for all DEGs from a list of DESeq2 results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PlotRNADEGPCAs.html">PlotRNADEGPCAs()</a></code> </p>
        </td>
        <td><p>Plot PCAs from variance stabilized counts for differentially expressed genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PlotRNAEDAPCAs.html">PlotRNAEDAPCAs()</a></code> </p>
        </td>
        <td><p>Plot PCAs from variance stabilized counts for EDA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PlotRNAHeatmaps.html">PlotRNAHeatmaps()</a></code> </p>
        </td>
        <td><p>Generate per comparison DEG heatmaps for a list of DESeq2 results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PlotRNASampleDistances.html">PlotRNASampleDistances()</a></code> </p>
        </td>
        <td><p>Visualize sample distances from variance stabilized counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PlotRNAVarianceTransformations.html">PlotRNAVarianceTransformations()</a></code> </p>
        </td>
        <td><p>Perform and visualize different variance stabilization transformations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PlotRNAVolcanoes.html">PlotRNAVolcanoes()</a></code> </p>
        </td>
        <td><p>Create volcano plots from a list of DESeq2 Results objects</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-chip-seq-analysis" class="hasAnchor"><a href="#section-chip-seq-analysis" class="anchor"></a>ChIP-seq analysis</h2>
          <p class="section-desc"><p>Functions that perform ChIPQC/DiffBind analyses, utilizing the utility and plotting functions to generate output tables and figures. These are the crux of the ChIP-seq portion of the package and what most users should use.</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="RunChIPQC.html">RunChIPQC()</a></code> </p>
        </td>
        <td><p>Run ChIPQC on a sample sheet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="RunDiffBind.html">RunDiffBind()</a></code> </p>
        </td>
        <td><p>Run DiffBind analysis on a samplesheet</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ProcessDBRs.html">ProcessDBRs()</a></code> </p>
        </td>
        <td><p>Extract, visualize, and save DBRs</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-chip-seq-plotting" class="hasAnchor"><a href="#section-chip-seq-plotting" class="anchor"></a>ChIP-seq Plotting</h2>
          <p class="section-desc"><p>Functions that generate plots for ChIP-seq data. Calling these individually can sometimes be useful.</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="PlotChIPAnnos.html">PlotChIPAnnos()</a></code> </p>
        </td>
        <td><p>Plot ChIP annotation breakdowns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PlotChIPHeatmaps.html">PlotChIPHeatmaps()</a></code> </p>
        </td>
        <td><p>Plot ChIP heatmaps</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PlotChIPPCAs.html">PlotChIPPCAs()</a></code> </p>
        </td>
        <td><p>Plot ChIP PCAs</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-enrichment-analyses" class="hasAnchor"><a href="#section-enrichment-analyses" class="anchor"></a>Enrichment Analyses</h2>
          <p class="section-desc"><p>This function performs GO/pathway enrichment analysis for lists of genes resulting from both RNA and ChIP-seq analyses.</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="PlotEnrichments.html">PlotEnrichments()</a></code> </p>
        </td>
        <td><p>Perform GO/Pathway enrichment for DEGs or DBRs via enrichR</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-internal-utilities" class="hasAnchor"><a href="#section-internal-utilities" class="anchor"></a>Internal utilities</h2>
          <p class="section-desc"><p>Functions used to create directory structure and save output. Users shouldnâ€™t need to call these explicitely.</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="CreateOutputStructure.html">CreateOutputStructure()</a></code> </p>
        </td>
        <td><p>Create output directory structure</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="SaveResults.html">SaveResults()</a></code> </p>
        </td>
        <td><p>Save results</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-rna-seq-analysis">RNA-seq analysis</a></li>
      <li><a href="#section-rna-seq-plotting">RNA-seq Plotting</a></li>
      <li><a href="#section-chip-seq-analysis">ChIP-seq analysis</a></li>
      <li><a href="#section-chip-seq-plotting">ChIP-seq Plotting</a></li>
      <li><a href="#section-enrichment-analyses">Enrichment Analyses</a></li>
      <li><a href="#section-internal-utilities">Internal utilities</a></li>
    </ul>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://j-andrews7.github.io'>Jared Andrews</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


