<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Extract, visualize, and save DBRs — ProcessDBRs • OneLinerOmics</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Extract, visualize, and save DBRs — ProcessDBRs" />
<meta property="og:description" content="ProcessDBRs wraps several plotting functions to generate figures 
specifically for differentially bound regions in a given comparison." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">OneLinerOmics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/ChIPseqAnalysis.html">Analyzing ChIP-seq data</a>
    </li>
    <li>
      <a href="../articles/RNAseqAnalysis.html">Analyzing RNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/j-andrews7/OneLinerOmics">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Extract, visualize, and save DBRs</h1>
    <small class="dont-index">Source: <a href='https://github.com/j-andrews7/OneLinerOmics/blob/master/R/ChIPseq.R'><code>R/ChIPseq.R</code></a></small>
    <div class="hidden name"><code>ProcessDBRs.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>ProcessDBRs</code> wraps several plotting functions to generate figures 
specifically for differentially bound regions in a given comparison.</p>
    </div>

    <pre class="usage"><span class='fu'>ProcessDBRs</span>(<span class='no'>results</span>, <span class='no'>outpath</span>, <span class='no'>txdb</span>, <span class='kw'>fdr.thresh</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>fc.thresh</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>promoters</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(-<span class='fl'>2000</span>, <span class='fl'>2000</span>), <span class='kw'>breaks</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span>(-<span class='fl'>3</span>,
  -<span class='fl'>1.0001</span>, <span class='kw'>length</span> <span class='kw'>=</span> <span class='fl'>250</span>), <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span>(-<span class='fl'>1</span>, -<span class='fl'>0.1</span>, <span class='kw'>length</span> <span class='kw'>=</span> <span class='fl'>250</span>), <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span>(-<span class='fl'>0.0999</span>, <span class='fl'>0.0999</span>,
  <span class='kw'>length</span> <span class='kw'>=</span> <span class='fl'>1</span>), <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span>(<span class='fl'>0.1</span>, <span class='fl'>1</span>, <span class='kw'>length</span> <span class='kw'>=</span> <span class='fl'>250</span>), <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span>(<span class='fl'>1.0001</span>, <span class='fl'>3</span>, <span class='kw'>length</span> <span class='kw'>=</span> <span class='fl'>250</span>)),
  <span class='kw'>heatmap.colors</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>heatmap.preset</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>reverse</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>plot.enrich</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>enrich.libs</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"GO_Molecular_Function_2018"</span>,
  <span class='st'>"GO_Cellular_Component_2018"</span>, <span class='st'>"GO_Biological_Process_2018"</span>,
  <span class='st'>"KEGG_2019_Human"</span>, <span class='st'>"Reactome_2016"</span>, <span class='st'>"BioCarta_2016"</span>, <span class='st'>"Panther_2016"</span>),
  <span class='kw'>flank.anno</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>flank.dist</span> <span class='kw'>=</span> <span class='fl'>5000</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>results</th>
      <td><p><code>DBA</code> object as returned by 
<code><a href='https://rdrr.io/pkg/DiffBind/man/dba.analyze.html'>dba.analyze</a></code>.</p></td>
    </tr>
    <tr>
      <th>outpath</th>
      <td><p>Path to directory to be used for output.</p></td>
    </tr>
    <tr>
      <th>txdb</th>
      <td><p><code>TxDb</code> object to use for annotation.</p></td>
    </tr>
    <tr>
      <th>fdr.thresh</th>
      <td><p>Number or numeric scalar indicating the false discovery 
rate (FDR) cutoff(s) to be used for determining "significant" differential 
binding. If multiple are given, multiple tables/plots will be generated 
using all combinations of <code>fdr.thresh</code> and <code>fc.thresh</code>.</p></td>
    </tr>
    <tr>
      <th>fc.thresh</th>
      <td><p>Number or numeric scalar indicating the log2 fold-change 
cutoff(s) to be used for determining "significant" differential binding.
If multiple are given, multiple tables/plots will be generated using all 
combinations of <code>padj.thresh</code> and <code>fc.thresh</code>.</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>Method used for differential binding e.g. <code>DBA_DESEQ2</code>.</p>
<p>Do not quote this parameter, it is a global variable from <code>DiffBind</code>.
  If a block was used, be sure to include that 
  (e.g. <code>DBA_DESEQ2_BLOCK</code>).</p></td>
    </tr>
    <tr>
      <th>promoters</th>
      <td><p>Scalar vector containing how many basepairs up and 
downstream of the TSS should be used to define gene promoters.</p></td>
    </tr>
    <tr>
      <th>breaks</th>
      <td><p>Vector of sequences to be used as breaks for signal heatmaps.</p></td>
    </tr>
    <tr>
      <th>heatmap.colors</th>
      <td><p>Character vector containing custom colors to use for 
heatmaps in hex (e.g. <code><a href='https://rdrr.io/r/base/c.html'>c("#053061", "#f5f5f5", "#67001f")</a></code>).</p></td>
    </tr>
    <tr>
      <th>heatmap.preset</th>
      <td><p>String indicating which of the color presets to use in
  heatmaps.</p>
<p>Available presets (low to high) are:</p><ul>
<li><p>"BuRd" Blue to red.</p></li>
<li><p>"OrPu" Orange to purple.</p></li>
<li><p>"BrTe" Brown to teal.</p></li>
<li><p>"PuGr" Purple to green.</p></li>
<li><p>"BuOr" Sea blue to orange.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>reverse</th>
      <td><p>Boolean indicating whether to flip heatmap color scheme 
(high color will become low, etc).</p></td>
    </tr>
    <tr>
      <th>plot.enrich</th>
      <td><p>Boolean indicating whether enrichment analyses for DBRs 
should be run and plotted for each comparison.</p></td>
    </tr>
    <tr>
      <th>enrich.libs</th>
      <td><p>A vector of valid <code>enrichR</code> libraries to test the 
genes against.</p></td>
    </tr>
    <tr>
      <th>flank.anno</th>
      <td><p>Boolean indicating whether flanking gene information for 
each peak should be retrieved. Useful for broad peaks and super enhancers.</p></td>
    </tr>
    <tr>
      <th>flank.distance</th>
      <td><p>Integer for distance from edges of peak to search for
  flanking genes. Ignored if <code>flank.anno = FALSE</code>.</p>
<p>Available libraries can be viewed with 
  <code><a href='https://rdrr.io/pkg/enrichR/man/listEnrichrDbs.html'>listEnrichrDbs</a></code> from the <code>enrichR</code> package.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>ProcessDBRs</code> is called by <a href='RunDiffBind.html'>RunDiffBind</a> but can also be re-run 
with the <a href='RunDiffBind.html'>RunDiffBind</a> output if you want to save time and don't need to 
generate all of the EDA/consensus figures again.</p>
<p>This function will generate many figures in addition to saving the results as
tables.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='RunDiffBind.html'>RunDiffBind</a></code>, for generating input for this function.</p></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#see-also">See also</a></li>
    </ul>

    <h2>Author</h2>
    <p>Jared Andrews</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://j-andrews7.github.io'>Jared Andrews</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


