<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Run DESeq2 analyses — RunDESeq2 • OneLinerOmics</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Run DESeq2 analyses — RunDESeq2" />
<meta property="og:description" content="RunDESeq2 performs a high-level differential gene expression analysis 
with DESeq2. It, along with ProcessDEGs, form the crux of the
RNA-seq portion of this package." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">OneLinerOmics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/ChIPseqAnalysis.html">Analyzing ChIP-seq data</a>
    </li>
    <li>
      <a href="../articles/RNAseqAnalysis.html">Analyzing RNA-seq data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/j-andrews7/OneLinerOmics">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run DESeq2 analyses</h1>
    <small class="dont-index">Source: <a href='https://github.com/j-andrews7/OneLinerOmics/blob/master/R/RNAseq.R'><code>R/RNAseq.R</code></a></small>
    <div class="hidden name"><code>RunDESeq2.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>RunDESeq2</code> performs a high-level differential gene expression analysis 
with <code>DESeq2</code>. It, along with <a href='ProcessDEGs.html'>ProcessDEGs</a>, form the crux of the
RNA-seq portion of this package.</p>
    </div>

    <pre class="usage"><span class='fu'>RunDESeq2</span>(<span class='no'>outpath</span>, <span class='no'>quants.path</span>, <span class='no'>samplesheet</span>, <span class='no'>tx2gene</span>, <span class='no'>level</span>,
  <span class='kw'>padj.thresh</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>fc.thresh</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>plot.annos</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>plot.box</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>plot.enrich</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>enrich.libs</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"GO_Molecular_Function_2018"</span>,
  <span class='st'>"GO_Cellular_Component_2018"</span>, <span class='st'>"GO_Biological_Process_2018"</span>,
  <span class='st'>"KEGG_2019_Human"</span>, <span class='st'>"Reactome_2016"</span>, <span class='st'>"BioCarta_2016"</span>, <span class='st'>"Panther_2016"</span>),
  <span class='kw'>top.n</span> <span class='kw'>=</span> <span class='fl'>100</span>, <span class='kw'>block</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>count.filt</span> <span class='kw'>=</span> <span class='fl'>10</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>outpath</th>
      <td><p>Path to directory to be used for output. Additional 
directories will be generated within this folder.</p></td>
    </tr>
    <tr>
      <th>quants.path</th>
      <td><p>Path to directory containing a directory for each sample
with a salmon-generated <code>quant.sf</code> file inside.</p></td>
    </tr>
    <tr>
      <th>samplesheet</th>
      <td><p>Path to samplesheet containing sample metadata.</p></td>
    </tr>
    <tr>
      <th>tx2gene</th>
      <td><p>Path to file with transcript IDs in first column and gene 
identifiers in second. Used for gene-level summarization.</p></td>
    </tr>
    <tr>
      <th>level</th>
      <td><p>String defining variable of interest.</p></td>
    </tr>
    <tr>
      <th>padj.thresh</th>
      <td><p>Number or numeric scalar indicating the adjusted p-value 
cutoff(s) to be used for determining "significant" differential expression.
If multiple are given, multiple tables/plots will be generated using all 
combinations of <code>padj.thresh</code> and <code>fc.thresh</code>.</p></td>
    </tr>
    <tr>
      <th>fc.thresh</th>
      <td><p>Number or numeric scalar indicating the log2 fold-change 
cutoff(s) to be used for determining "significant" differential expression.
If multiple are given, multiple tables/plots will be generated using all 
combinations of <code>padj.thresh</code> and <code>fc.thresh</code>.</p></td>
    </tr>
    <tr>
      <th>plot.annos</th>
      <td><p>String or character vector defining the column(s) in 
<code>samplesheet</code> to use to annotate figures.</p></td>
    </tr>
    <tr>
      <th>plot.box</th>
      <td><p>Boolean indicating whether box plots for DEGs should be 
created for each comparison. If so, the <code>top.n</code> genes will be plotted.
This step is quite time-consuming with many genes.</p></td>
    </tr>
    <tr>
      <th>plot.enrich</th>
      <td><p>Boolean indicating whether enrichment analyses for DEGs 
should be run and plotted for each comparison.</p></td>
    </tr>
    <tr>
      <th>enrich.libs</th>
      <td><p>Vector of valid <code>enrichR</code> libraries to test the 
  genes against.</p>
<p>Available libraries can be viewed with <code>listEnrichrDbs</code> from the
  <code>enrichR</code> package.</p></td>
    </tr>
    <tr>
      <th>top.n</th>
      <td><p>Number of differentially expressed genes to create boxplots for, 
ranked by adj. p-value after applying <code>padj.thresh</code> and 
<code>fc.thresh</code> thresholds. If multiple thresholds are provided, the 
lowest fold-change and highest adj. p-value thresholds will be used.</p></td>
    </tr>
    <tr>
      <th>block</th>
      <td><p>String or character vector defining the column(s) in 
<code>samplesheet</code> to use to block for unwanted variance, e.g. batch or 
technical effects.</p></td>
    </tr>
    <tr>
      <th>count.filt</th>
      <td><p>Number indicating read threshold. Genes with fewer counts 
than this number summed across all samples will be removed from the
analysis.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Named List containing a list named 'res.list' containing 
  DESeqResults objects for all comparisons generated by 
  <code><a href='ProcessDEGs.html'>ProcessDEGs</a></code>, a DESeqDataSet object named 'dds'
  from running <code><a href='https://rdrr.io/pkg/DESeq2/man/DESeq.html'>DESeq</a></code>, a 
  RangedSummarizedExperiment object named 'rld' from running 
  <code><a href='https://rdrr.io/pkg/DESeq2/man/rlog.html'>rlog</a></code>, and a RangedSummarizedExperiment 
  object from running <code><a href='https://rdrr.io/pkg/DESeq2/man/vst.html'>vst</a></code> named 'vsd'.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The default parameters should be an adequate starting place for most users,
but lazy folks can provide multiple thresholds to <code>padj.thresh</code>
and/or <code>fc.thresh</code> if they aren't sure how stringent or lenient they 
need to be with their data.</p>
<p>It's generally best to provide an empty directory as the output path, as
several directories will be generated.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='https://rdrr.io/pkg/DESeq2/man/DESeq.html'>DESeq</a></code>, for more about differential expression analysis.
<code><a href='ProcessDEGs.html'>ProcessDEGs</a></code>, for analyzing and visualizing the results.</p></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#see-also">See also</a></li>
    </ul>

    <h2>Author</h2>
    <p>Jared Andrews</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://j-andrews7.github.io'>Jared Andrews</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


