#' Run DiffBind analysis on a samplesheet
#'
#' \code{RunDiffBind} performs a high-level differential binding analysis 
#' with \code{DiffBind}. It, along with \link{ProcessDBRs}, form the crux of the
#' ChIP-seq portion of this package. 
#'
#' The default parameters should be an adequate starting place for most users,
#' but lazy folks can provide multiple thresholds to \code{padj.thresh}
#' and/or \code{fc.thresh} if they aren't sure how stringent or lenient they 
#' need to be with their data.
#'
#' It's generally best to provide an empty directory as the output path, as
#' several directories will be generated. 
#'
#' 
#'
#' @param outpath Path to directory to be used for output. Additional 
#'   directories will be generated within this folder.
#' @param samplesheet Path to samplesheet containing sample metadata.
#' @param level String defining variable of interest from \code{samplesheet}. 
#'   Must be one of: "Treatment", "Condition", "Tissue", or "Factor".
#' @param se Path to file containing consensus SEs, which will be used to
#'   to annotate whether individual peaks fall within an SE or not.
#' @param fdr.thresh Number or numeric scalar indicating the false discovery 
#'   rate (FDR) cutoff(s) to be used for determining "significant" differential 
#'   binding. If multiple are given, multiple tables/plots will be generated 
#'   using all combinations of \code{fdr.thresh} and \code{fc.thresh}.
#' @param fc.thresh Number or numeric scalar indicating the log2 fold-change 
#'   cutoff(s) to be used for determining "significant" differential binding.
#'   If multiple are given, multiple tables/plots will be generated using all 
#'   combinations of \code{padj.thresh} and \code{fc.thresh}.
#' @param block String or character vector defining the column(s) in 
#'   \code{samplesheet} to use to block for unwanted variance, e.g. batch or 
#'   technical effects. Must be one of: "Treatment", "Condition", "Tissue", 
#'   or "Factor".
#' @param n.consensus Number of samples in which peaks must overlap for the 
#'   peaks to be merged and included in the consensus peak set. 
#' @param breaks Vector of sequences to be used as breaks for signal heatmaps.
#' @param heatmap.colors Character vector containing custom colors to use for 
#'   heatmaps in hex (e.g. \code{c("#053061", "#f5f5f5", "#67001f")}).
#' @param heatmap.preset String indicating which of the color presets to use in
#'   heatmaps.
#'
#'   Available presets (low to high) are:
#'   \itemize{
#'      \item "BuRd" Blue to red.
#'      \item "OrPu" Orange to purple. 
#'      \item "BrTe" Brown to teal.
#'      \item "PuGr" Purple to green.
#'   }
#' @param plot.enrich Boolean indicating whether enrichment analyses for DBRs 
#'   should be run and plotted for each comparison. 
#' @param enrich.libs Vector of valid \code{enrichR} libraries to test the 
#'   genes against.
#'   
#'
#'   Available libraries can be viewed with 
#'   \code{\link[enrichR]{listEnrichrDbs}} from the \code{enrichR} package.
#' @param promoters Vector containing how many basepairs up and downstream of 
#'   the TSS should be used to define gene promoters.
#' @param method String indicating method to be used for differential expression 
#'   analysis. Can be "DESeq2" or "edgeR".
#' @return Named List containing a list named 'res.list' containing 
#'   \linkS4class{DESeqResults} objects for all comparisons generated by 
#'   \code{\link{ProcessDEGs}}, a \linkS4class{DESeqDataSet} object named 'dds'
#'   from running \code{\link[DESeq2]{DESeq}}, a 
#'   \linkS4class{RangedSummarizedExperiment} object named 'rld' from running 
#'   \code{\link[DESeq2]{rlog}}, and a \linkS4class{RangedSummarizedExperiment} 
#'   object from running \code{\link[DESeq2]{vst}} named 'vsd'.
#'
#' @importFrom utils read.table
#' @importFrom TxDb.Hsapiens.UCSC.hg19.knownGene 
#'   TxDb.Hsapiens.UCSC.hg19.knownGene
#' @import DiffBind
#'
#' @export
#'
#' @author Jared Andrews
#'
#' @seealso
#' \code{\link[DiffBind]{dba}}, \code{\link[DiffBind]{dba.count}}, 
#' \code{\link[DiffBind]{dba.contrast}}, \code{\link[DiffBind]{dba.analyze}},
#' \code{\link[DiffBind]{dba.report}} for more about ChIP-seq differential
#' binding analysis.
#'
#' \code{\link{ProcessDBRs}}, for analyzing and visualizing the results.
#'
RunDiffBind <- function(outpath, samplesheet, 
  level = c("Treatment", "Condition", "Tissue", "Factor"), 
  se = NULL, 
  fdr.thresh = 0.05, 
  fc.thresh = 2, 
  block = NULL, 
  heatmap.colors = NULL, 
  heatmap.preset = NULL, 
  n.consensus = 2, 
  breaks = c(seq(-3, -1, length = 250), seq(-1, -0.1, length = 250),
    seq(-0.1, 0.1,length = 1), seq(0.1, 1, length = 250), 
    seq(1, 3, length = 250)),
  plot.enrich = TRUE, 
  enrich.libs = c("GO_Molecular_Function_2018", 
  "GO_Cellular_Component_2018", "GO_Biological_Process_2018", "KEGG_2019_Human",
  "Reactome_2016", "BioCarta_2016", "Panther_2016"),
  promoters = c(-2000, 2000),
  method = c("DESeq2", "edgeR")) {

  message("### SETUP ###\n")
  rblock = NULL

  ### CHECKING ARGUMENTS ###
  message("\n# CHECKING ARGUMENTS #\n")
  level <- match.arg(level)
  if (level == "Treatment") {
    rlevel <- DBA_TREATMENT
  } else if (level == "Condition") {
    rlevel <- DBA_CONDITION
  } else if (level == "Tissue") {
    rlevel <- DBA_TISSUE
  } else if (level == "Factor") {
    rlevel <- DBA_FACTOR
  }

  if (!is.null(block)) {
    block <- match.arg(block, choices = c("Treatment", "Condition", "Tissue", 
      "Factor"))

    if (block == "Treatment") {
      rblock <- DBA_TREATMENT
    } else if (block == "Condition") {
      rblock <- DBA_CONDITION
    } else if (block == "Tissue") {
      rblock <- DBA_TISSUE
    } else if (block == "Factor") {
      rblock <- DBA_FACTOR
    }
  }

  method <- match.arg(method)
  if (!is.null(rblock)) {
    if (method == "DESeq2") {
      method <- DBA_DESEQ2_BLOCK
    } else {
      method <- DBA_EDGER_BLOCK
    }
  } else {
    if (method == "DESeq2") {
      method <- DBA_DESEQ2
    } else {
      method <- DBA_EDGER
    }
  }

  # Set up colors/breaks for heatmaps later.
  hmap.colors <- .heatmap_colors(breaks = breaks, preset = heatmap.preset, 
    custom.colors = heatmap.colors)
  
  ### CREATE DIRECTORY STRUCTURE ###
  message("# CREATING DIRECTORY STRUCTURE #\n\n")
  setup <- CreateOutputStructure(block, level, base = outpath, chip = TRUE)
  base <- setup$base
  
  ### LOAD SUPER ENHANCERS ###
  if (!is.null(se)) {
    message("\n # LOADING SUPER ENHANCERS #\n")
    se <- read.table(se, sep = '\t', header = TRUE)
  }  

  ### FINDING DIFFERENTIALLY BOUND REGIONS ###
  message("### FINDING DIFFERENTIALLY BOUND REGIONS ###\n\n")
  samps <- dba(sampleSheet = samplesheet, minOverlap = n.consensus)
  count <- dba.count(samps, minOverlap = n.consensus)
  cont <- dba.contrast(count, categories = level, block = rblock)
  results <- dba.analyze(cont)
  
  # CONSENSUS PEAKS #
  message('# DEALING WITH CONSENSUS PEAKS #\n\n')
  txdb=TxDb.Hsapiens.UCSC.hg19.knownGene
  if (!is.null(rblock)) {
    report <- dba.report(results, th = 1, bCalled = TRUE, 
      bCalledDetail = TRUE, bCounts = TRUE, method = method)
  } else {
    report <- dba.report(results, th = 1, bCalled = TRUE, 
      bCalledDetail = TRUE, bCounts = TRUE, method = method)
  }

  # ANNOTATION #
  message('# ANNOTATING & MAKING CONSENSUS PLOTS #\n\n')
  peakAnno <- annotatePeak(report, tssRegion = promoters,
                       TxDb = txdb, annoDb = "org.Hs.eg.db")
  
  df <- data.frame(peakAnno)

  pdf(paste0(base, '/', mark, ".ConsensusPeaks.Figures.pdf"))

  plotAnnoPie(peakAnno)
  plotAnnoBar(peakAnno)
  vennpie(peakAnno)
  upsetplot(peakAnno)
  plotDistToTSS(peakAnno, title = "Distribution of Peaks Relative to TSS")
  dba.plotPCA(results, th = 1, sub = paste0(
    "Consensus Peaks"), method = method)

  dba.plotHeatmap(results, report = report,
    contrast = 1, margin = 15, correlations = FALSE, scale = "row", 
    density.info = "none", colScheme = hmap.colors, breaks = breaks, 
    main = paste0("Consensus Peaks\n", "Top 1000 (by variability)"))
  dba.plotHeatmap(results, report = report, contrast = 1, 
    margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
    colScheme = hmap.colors, breaks = breaks, Colv = NULL, 
    main = paste0("Consensus Peaks\n", "Top 1000 (by variability)"))
  dba.plotHeatmap(results, report = report, contrast = 1, 
    margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
    colScheme = hmap.colors, breaks = breaks, maxSites = 10000, 
    main = paste0("Consensus Peaks\n", "Top 10000 (by variability)"))
  dba.plotHeatmap(results, report = report, contrast = 1, 
    margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
    colScheme = hmap.colors, breaks = breaks, maxSites = 10000, Colv = NULL, 
    main = paste0("Consensus Peaks\n", "Top 10000 (by variability)"))
  dba.plotHeatmap(results, report = report, contrast = 1, 
    margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
    colScheme = hmap.colors, breaks = breaks, maxSites = 10000, 
    main = paste0("Consensus Peaks\n", "All Peaks"))
  dba.plotHeatmap(results, report = report, contrast = 1, 
    margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
    colScheme = hmap.colors, breaks = breaks, maxSites = 10000, Colv = NULL, 
    main = paste0("Consensus Peaks\n", "All Peaks"))
  dba.plotHeatmap(results, report = report, contrast = 1, 
    margin = 15, density.info = "none", main = paste0("Consensus Peaks\n", 
      "Correlation"))

  dev.off()
  
  # DB PEAKS #
  message('# DEALING WITH DB PEAKS #\n\n')
  # Subset so that we can compare localization and pathways between these groups as well.
  if (!is.null(rblock)){
    reportdb = dba.report(results, th = 0.05, bCalled = TRUE, 
      bCalledDetail = TRUE, bCounts = TRUE, method = method)
  } else {
    reportdb = dba.report(results, th = 0.05, bCalled = TRUE, 
      bCalledDetail = TRUE, bCounts = TRUE, method = method)
  }
  
  g1up = reportdb[(reportdb$Fold > 0)]
  g2up = reportdb[(reportdb$Fold < 0)]

  # Create a named list of our subsets.
  files = GRangesList(reportdb, g1up, g2up)
  names(files) = c("All_DB", paste0(g1, ".up"), paste0(g2, ".up"))
  peakAnnoList <- lapply(files, annotatePeak, TxDb = txdb,
    tssRegion=c(-2000, 2000), verbose = FALSE, annoDb = "org.Hs.eg.db")
  ProcessDBRs(peakAnnoList, reportdb, results, rblock, g1, g2, base, mark, 
    color, breaks)
  
  # Categorize and save all peaks.
  dfs <- list("peaks" = df, "ses" = se)
  x <- .categorize_peaks(dfs)
  
  write.table(x, file = paste0(base, '/', g1, '-v-', g2, ".", mark, 
    ".ConsensusPeaks.txt"), sep = "\t", quote = FALSE, row.names = FALSE)
  
  x.db <- x[x$FDR <= 0.05,]
  write.table(x.db, file = paste0(base, '/', g1, '-v-', g2, ".", mark, 
    ".DBRs.FDR0.05.txt"), sep = "\t", quote = FALSE, row.names = FALSE)
  
  return(x)
}


#' Extract, visualize, and save DBRs
#'
#' \code{ProcessDBRs} wraps several plotting functions to generate figures 
#' specifically for the differentially bound regions between each possible
#' comparison. 
#' 
#' \code{ProcessDBRs} is called by \link{RunDiffBind} but can also be re-run 
#' with the \link{RunDiffBind} output or its own output if you want to save time 
#' and don't need to generate all of the exploratory data analysis figures 
#' again.
#'
#' This function will generate many figures in addition to saving the results as
#' tables.
#'
#' @param results A \linkS4class{DESeqDataSet} object as returned by 
#'   \code{\link[DESeq2]{DESeq}}.
#' @param rblock Boolean indicating whether a blocking variable was used or not.
#' @param outpath Path to directory to be used for output.
#' @param level String defining variable of interest in samplesheet.
#'   Must be one of: "Treatment", "Condition", "Tissue", or "Factor".
#' @param fdr.thresh Number or numeric scalar indicating the false discovery 
#'   rate (FDR) cutoff(s) to be used for determining "significant" differential 
#'   binding. If multiple are given, multiple tables/plots will be generated 
#'   using all combinations of \code{fdr.thresh} and \code{fc.thresh}.
#' @param fc.thresh Number or numeric scalar indicating the log2 fold-change 
#'   cutoff(s) to be used for determining "significant" differential binding.
#'   If multiple are given, multiple tables/plots will be generated using all 
#'   combinations of \code{padj.thresh} and \code{fc.thresh}.
#' @param breaks Vector of sequences to be used as breaks for signal heatmaps.
#' @param heatmap.colors Character vector containing custom colors to use for 
#'   heatmaps in hex (e.g. \code{c("#053061", "#f5f5f5", "#67001f")}).
#' @param heatmap.preset String indicating which of the color presets to use in
#'   heatmaps.
#'
#'   Available presets (low to high) are:
#'   \itemize{
#'      \item "BuRd" Blue to red.
#'      \item "OrPu" Orange to purple. 
#'      \item "BrTe" Brown to teal.
#'      \item "PuGr" Purple to green.
#'   }
#' @param plot.enrich Boolean indicating whether enrichment analyses for DBRs 
#'   should be run and plotted for each comparison.
#' @param enrich.libs A vector of valid \code{enrichR} libraries to test the 
#'   genes against.
#'
#'   Available libraries can be viewed with 
#'   \code{\link[enrichR]{listEnrichrDbs}} from the \code{enrichR} package.
#' @return Named List containing a list named 'res.list' containing 
#'   \linkS4class{DESeqResults} objects for all comparisons generated by 
#'   \code{\link{ProcessDEGs}}, a \linkS4class{DESeqDataSet} object named 'dds'
#'   from running \code{\link[DESeq2]{DESeq}}, a 
#'   \linkS4class{RangedSummarizedExperiment} object named 'rld' from running 
#'   \code{\link[DESeq2]{rlog}}, and a \linkS4class{RangedSummarizedExperiment} 
#'   object from running \code{\link[DESeq2]{vst}} named 'vsd'.
#'
#' @import DiffBind
#'
#' @export
#'
#' @author Jared Andrews
#'
#' @seealso
#' \code{\link{RunDiffBind}}, for generating input for this function.
#'
ProcessDBRs <- function(results, rblock, outpath, level, 
  fdr.thresh = 0.05, 
  fc.thresh = 2, 
  breaks = c(seq(-3, -1, length = 250), seq(-1, -0.1, length = 250),
    seq(-0.1, 0.1,length = 1), seq(0.1, 1, length = 250), 
    seq(1, 3, length = 250)), 
  heatmap.colors = NULL, 
  heatmap.preset = NULL, 
  plot.enrich = TRUE, 
  enrich.libs = c("GO_Molecular_Function_2018", 
  "GO_Cellular_Component_2018", "GO_Biological_Process_2018", "KEGG_2019_Human",
  "Reactome_2016", "BioCarta_2016", "Panther_2016")) {

  pdf(paste0(base ,"/", g1, '-v-', g2, ".", mark, ".DBRs.Figures.pdf"))

  if (!is.null(rblock)) {
    plotAnnoBar(peakAnnoList)
    plotDistToTSS(peakAnnoList, 
      title = "Distribution of Regions Relative to TSS")
    dba.plotPCA(results, th = 0.05, contrast = 1, 
      sub = paste0(g1, '-v-', g2, " DBRs\n", mark), method = DBA_DESEQ2_BLOCK)
    dba.plotMA(results, th = 0.05, contrast = 1, method = DBA_DESEQ2_BLOCK)
    dba.plotMA(results, th = 0.05, contrast = 1, bXY = TRUE, 
      method = DBA_DESEQ2_BLOCK)
    dba.plotVolcano(results, th = 0.05, contrast = 1, method = DBA_DESEQ2_BLOCK)
    dba.plotBox(results, th = 0.05, contrast = 1, method = DBA_DESEQ2_BLOCK)

    dba.plotHeatmap(results, method = DBA_DESEQ2_BLOCK, th = 0.05, contrast = 1, 
      margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
      colScheme = color, breaks = breaks, main = paste0(g1, '-v-', g2, 
        " DBRs\n", mark, " Top 1000"))
    dba.plotHeatmap(results, method = DBA_DESEQ2_BLOCK, th = 0.05, contrast = 1, 
      margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
      colScheme = color, breaks = breaks, Colv = NULL, 
      main = paste0(g1, '-v-', g2, " DBRs\n", mark, " Top 1000"))
    dba.plotHeatmap(results, method = DBA_DESEQ2_BLOCK, th = 0.05, contrast = 1, 
      margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
      colScheme = color, breaks = breaks, maxSites = 20000, 
      main = paste0(g1, '-v-', g2, " DBRs\n", mark, " Top 20000"))
    dba.plotHeatmap(results, method = DBA_DESEQ2_BLOCK, th = 0.05, contrast = 1, 
      margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
      colScheme = color, breaks = breaks, maxSites = 20000, Colv = NULL, 
      main = paste0(g1, '-v-', g2, " DBRs\n", mark, " Top 20000"))
    dba.plotHeatmap(results, method = DBA_DESEQ2_BLOCK, th = 0.05, contrast = 1, 
      margin = 15, density.info = "none", 
      main = paste0(g1, "-v-", g1, " DBRs\n", mark, " - Correlation"))
  } else {
    plotAnnoBar(peakAnnoList)
    plotDistToTSS(peakAnnoList, 
      title = "Distribution of Regions Relative to TSS")
    dba.plotPCA(results, th = 0.05, contrast = 1, 
      sub = paste0(g1, '-v-', g2, " DBRs\n", mark), method = DBA_DESEQ2)
    dba.plotMA(results, th = 0.05, contrast = 1, method = DBA_DESEQ2)
    dba.plotMA(results, th = 0.05, contrast = 1, bXY = TRUE, 
      method = DBA_DESEQ2)
    dba.plotVolcano(results, th = 0.05, contrast = 1, method = DBA_DESEQ2)
    dba.plotBox(results, th = 0.05, contrast = 1, method = DBA_DESEQ2)

    dba.plotHeatmap(results, method = DBA_DESEQ2, th = 0.05, contrast = 1, 
      margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
      colScheme = color, breaks = breaks, 
      main = paste0(g1, '-v-', g2, " DBRs\n", mark, " Top 1000"))
    dba.plotHeatmap(results, method = DBA_DESEQ2, th = 0.05, contrast = 1, 
      margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
      colScheme = color, breaks = breaks, Colv = NULL, 
      main = paste0(g1, '-v-', g2, " DBRs\n", mark, " Top 1000"))
    dba.plotHeatmap(results, method = DBA_DESEQ2, th = 0.05, contrast = 1, 
      margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
      colScheme = color, breaks = breaks, maxSites = 20000, 
      main = paste0(g1, '-v-', g2, " DBRs\n", mark, " Top 20000"))
    dba.plotHeatmap(results, method = DBA_DESEQ2, th = 0.05, contrast = 1, 
      margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
      colScheme = color, breaks = breaks, maxSites = 20000, Colv = NULL, 
      main = paste0(g1, '-v-', g2, " DBRs\n", mark, " Top 20000"))
    dba.plotHeatmap(results, method = DBA_DESEQ2, th = 0.05, contrast = 1, 
      margin = 15, density.info = "none", 
      main = paste0(g1, "-v-", g1, " DBRs\n", mark, " - Correlation"))
  }
  dev.off()
  
  RunEnrichments(peakAnnoList, g1, g2, base)
}


ProcessSEs <- function(samplesheet, rblock, g1, g2, base, breaks, color) {
  # samplesheet = File path to SE samplesheet.
  # rblock = Blocking factor.
  # g1 = Name of group 1.
  # g2 = Name of group 2.
  # base = File path to output folder.
  # breaks = Breaks to be used for heatmaps.
  # color = Color scheme to be used for heatmaps.
  
  samps = dba(sampleSheet = samplesheet)
  
  count = dba.count(samps)
  if (!is.null(rblock)) {
    cont = dba.contrast(count, count$masks[[g1]], count$masks[[g2]], g1, g2, 
      block=rblock, minMembers = 2)
  } else {
    cont = dba.contrast(count, count$masks[[g1]], count$masks[[g2]], g1, g2, 
      minMembers = 2)
  }
  results = dba.analyze(cont)
  
  # CONSENSUS PEAKS #
  message('\n# DEALING WITH CONSENSUS SEs #\n\n')
  txdb = TxDb.Hsapiens.UCSC.hg19.knownGene
  if (!is.null(rblock)) {
    report = dba.report(results, th = 1, bCalled = TRUE, bCalledDetail = TRUE, 
      bCounts = TRUE, method = DBA_DESEQ2_BLOCK)
  } else {
    report = dba.report(results, th = 1, bCalled = TRUE, bCalledDetail = TRUE, 
      bCounts = TRUE, method = DBA_DESEQ2)
  }

  # ANNOTATION #
  message('\n# ANNOTATING & MAKING GENERIC PLOTS #\n\n')
  peakAnno <- annotatePeak(report, tssRegion = c(-2000, 2000),
                       TxDb = txdb, annoDb = "org.Hs.eg.db")
  
  df = data.frame(peakAnno)
  
  
  write.table(df, row.names = FALSE, quote = FALSE, sep = "\t", 
    file = paste0(base, "/", g1, '-v-', g2, ".", mark, ".Consensus.SEs.txt"))

  pdf(paste0(base, "/", g1, '-v-', g2, ".", mark, ".Consensus.SEs.Figures.pdf"))

  if (!is.null(rblock)) {
    plotAnnoPie(peakAnno)
    plotAnnoBar(peakAnno)
    vennpie(peakAnno)
    upsetplot(peakAnno)
    plotDistToTSS(peakAnno, title = "Distribution of SEs Relative to TSS")
    dba.plotPCA(results, report = report, 
      sub = paste0(g1, '-v-', g2, " Consensus\n", mark, " SEs"), 
      method = DBA_DESEQ2_BLOCK)
    
    dba.plotHeatmap(results, method = DBA_DESEQ2_BLOCK, report = report, 
      margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
      colScheme = color, breaks = breaks, 
      main = paste0(g1, '-v-', g2, " Consensus\n", mark, " Top 1000"))
    dba.plotHeatmap(results, method = DBA_DESEQ2_BLOCK, report = report, 
      margin = 15, correlations = FALSE, scale = "row", density.info = "none", 
      colScheme = color, breaks = breaks, Colv = NULL, 
      main = paste0(g1, '-v-', g2, " Consensus\n", mark, " Top 1000"))
    dba.plotHeatmap(results, method = DBA_DESEQ2_BLOCK, report = report, 
      margin = 15, density.info = "none", 
      main = paste0(g1, "-v-", g1, " Consensus\n", mark, " SEs - Correlation"))
  } else {
    plotAnnoPie(peakAnno)
    plotAnnoBar(peakAnno)
    vennpie(peakAnno)
    upsetplot(peakAnno)
    plotDistToTSS(peakAnno, title = "Distribution of SEs Relative to TSS")
    dba.plotPCA(results, report = report, 
      sub = paste0(g1, '-v-', g2, " Consensus\n", mark, " SEs"), 
      method = DBA_DESEQ2)
    
    dba.plotHeatmap(results, method = DBA_DESEQ2, report = report, margin = 15, 
      correlations = FALSE, scale = "row", density.info = "none", 
      colScheme = color, breaks = breaks, 
      main = paste0(g1, '-v-', g2, " Consensus\n", mark, " Top 1000"))
    dba.plotHeatmap(results, method = DBA_DESEQ2, report = report, margin = 15, 
      correlations = FALSE, scale = "row", density.info = "none", 
      colScheme = color, breaks = breaks, Colv = NULL, 
      main = paste0(g1, '-v-', g2, " Consensus\n", mark, " Top 1000"))
    dba.plotHeatmap(results, method = DBA_DESEQ2, report = report, margin = 15, 
      density.info = "none", 
      main = paste0(g1, "-v-", g1, " Consensus\n", mark, " SEs - Correlation"))
  }
  dev.off()
  
  return(df)
}


#' @importFrom GenomicRanges makeGRangesFromDataFrame findOverlaps
#' @importFrom S4Vectors queryHits
.categorize_peaks <- function(dfs, groups=FALSE) {
  # dfs = Named list of two dataframes - first for peaks, second for SEs.
  
  # Make them GRanges objects for easy overlap.
  peaks <- makeGRangesFromDataFrame(dfs$peaks, keep.extra.columns = TRUE)
  ses <- makeGRangesFromDataFrame(dfs$ses, keep.extra.columns = TRUE)
  
  peaks$inSE <- "FALSE"
  hits <- findOverlaps(peaks, ses)
  peaks$inSE[queryHits(hits)] <- "TRUE"
  
  # Add a column for the peak category.
  peaks$cat <- "PROMOTER"
  peaks$cat[((startsWith(peaks$annotation, "Promoter")) & 
    peaks$inSE=="TRUE")] <- "PROM.SE"
  peaks$cat[((startsWith(peaks$annotation, "Intron") | 
    startsWith(peaks$annotation, "Downstream") | 
    startsWith(peaks$annotation, "Distal") ) & peaks$inSE=="TRUE") ] <- "CE.SE"
  peaks$cat[((startsWith(peaks$annotation, "Intron") | 
    startsWith(peaks$annotation, "Downstream") | 
    startsWith(peaks$annotation, "Distal") ) & peaks$inSE=="FALSE") ] <- "CE"
  peaks$cat[((startsWith(peaks$annotation, "Exon")) & 
    peaks$inSE=="FALSE") ] <- "EXON"
  peaks$cat[((startsWith(peaks$annotation, "Exon")) & 
    peaks$inSE=="TRUE") ] <- "EXON.SE"
  
  
  # Rearrange some columns for easy plotting later.
  final <- as.data.frame(peaks)
  if (groups == FALSE){
    nsamps <- (ncol(final) - 27)/2
    final <- subset(final, select=c(1:13,(14+nsamps):ncol(final), 
      14:(ncol(final)-(14+nsamps))))
  }
  
  return(final)
}


RunEnrichments <- function(peakAnnoList, g1, g2, outpath) {
  message("# GO ENRICHMENT #\n\n")
  genes = lapply(peakAnnoList, function(i) as.data.frame(i)$geneId)
  genes2 = lapply(peakAnnoList[2:3], function(i) as.data.frame(i)$geneId)
}

